From 31b7c95a864c07256376998a9b96c0af20523c17 Mon Sep 17 00:00:00 2001
From: "40902469+Samsuik@users.noreply.github.com"
 <40902469+Samsuik@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:38:26 +0000
Subject: [PATCH] Smooth Potting


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index cd4810b80..29619c61c 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -275,4 +275,9 @@ public class BlossomConfig {
         hopperInterval = Math.max(getInt("items.hoppers.delay", 0), 0) + 1;
     }
 
+    public static boolean smoothPotting;
+    private static void SmoothPotting() {
+        smoothPotting = getBoolean("projectiles.potion.smooth", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index e7cca3c6c..b11a7f83a 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -131,6 +131,12 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
                     AxisAlignedBB axisalignedbb = entity1.getBoundingBox().grow((double) f, (double) f, (double) f);
                     MovingObjectPosition movingobjectposition1 = axisalignedbb.a(vec3d, vec3d1);
 
+                    // Blossom start - Smooth Potting
+                    if (this instanceof EntityPotion && me.samsuik.blossom.configuration.BlossomConfig.smoothPotting &&
+                        movingobjectposition1 == null && getBoundingBox().b(entity1.getBoundingBox())) {
+                        movingobjectposition1 = new MovingObjectPosition(entity1);
+                    }
+                    // Blossom end
                     if (movingobjectposition1 != null) {
                         double d1 = vec3d.distanceSquared(movingobjectposition1.pos);
 
-- 
2.49.0

