From bd53836f6abb6973b6a5192174ba36f726457788 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Wed, 21 Jun 2023 18:54:40 +0100
Subject: [PATCH] Use sane initial capacity in CraftWorld entity getters


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 958fe4ffb..44e55e19e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -644,7 +644,7 @@ public class CraftWorld implements World {
     }
 
     public List<Entity> getEntities() {
-        List<Entity> list = new ArrayList<Entity>();
+        List<Entity> list = new ArrayList<Entity>(world.entityList.size()); // Blosom - optimise bukkit entity getters
 
         for (Object o : world.entityList) {
             if (o instanceof net.minecraft.server.Entity) {
@@ -662,7 +662,7 @@ public class CraftWorld implements World {
     }
 
     public List<LivingEntity> getLivingEntities() {
-        List<LivingEntity> list = new ArrayList<LivingEntity>();
+        List<LivingEntity> list = new ArrayList<LivingEntity>((world.entityList.size() * 2) / 3); // Blosom - optimise bukkit entity getters
 
         for (Object o : world.entityList) {
             if (o instanceof net.minecraft.server.Entity) {
-- 
2.40.0.windows.1

