From 8f9cb0af0e7b911620f319bad8777f5c51f260c9 Mon Sep 17 00:00:00 2001
From: Hugo Manrique <hugmanrique@gmail.com>
Date: Mon, 16 Jul 2018 12:42:20 +0200
Subject: [PATCH] Avoid item merge if stack size above max stack size


diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 24707c805..8e01ccd26 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -155,6 +155,11 @@ public class EntityItem extends Entity implements HopperPusher {
     // Spigot end
 
     private void w() {
+        // Paper start - avoid item merge if stack size above max stack size
+        ItemStack stack = getItemStack();
+        if (stack.count >= stack.getMaxStackSize()) return;
+        // Paper end
+
         // Spigot start
         double radius = world.spigotConfig.itemMerge;
         Iterator iterator = this.world.a(EntityItem.class, this.getBoundingBox().grow(radius, radius, radius)).iterator();
-- 
2.50.1

