From 6591d07696438f1fcaa5cdd91d7114c0a1553f9b Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Mon, 8 Oct 2018 17:51:52 +0100
Subject: [PATCH] Backport: Prevent mob spawning from loading/generating chunks


diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
index ea17d7e6f..4b7f929e8 100644
--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
+++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
@@ -140,9 +140,12 @@ public final class SpawnerCreature {
                             int i2 = blockposition1.getX();
                             int j2 = blockposition1.getY();
                             int k2 = blockposition1.getZ();
-                            Block block = worldserver.getType(blockposition1).getBlock();
+                            // Paper start
+                            IBlockData iblockdata = worldserver.getTypeIfLoaded(blockposition1);
+                            Block block;
 
-                            if (!block.isOccluding()) {
+                            if (iblockdata != null && !(block = iblockdata.getBlock()).isOccluding()) {
+                                // Paper end
                                 int l2 = 0;
                                 int i3 = 0;
 
@@ -165,7 +168,7 @@ public final class SpawnerCreature {
                                                 float f = (float) j3 + 0.5F;
                                                 float f1 = (float) l3 + 0.5F;
 
-                                                if (!worldserver.isPlayerNearbyWhoAffectsSpawning((double) f, (double) k3, (double) f1, 24.0D) && blockposition.c((double) f, (double) k3, (double) f1) >= 576.0D) { // PaperSpigot - Affects Spawning API
+                                                if (worldserver.getChunkIfLoaded(blockposition) != null && !worldserver.isPlayerNearbyWhoAffectsSpawning((double) f, (double) k3, (double) f1, 24.0D) && blockposition.c((double) f, (double) k3, (double) f1) >= 576.0D) { // PaperSpigot - Affects Spawning API
                                                     if (biomebase_biomemeta == null) {
                                                         biomebase_biomemeta = worldserver.a(enumcreaturetype, blockposition2);
                                                         if (biomebase_biomemeta == null) {
-- 
2.40.0.windows.1

