From 40ea2ebb4ddb0e112ef81e193da072bfdebd0be8 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sun, 14 Apr 2019 20:59:16 +0100
Subject: [PATCH] Don't drop items into the world if BlockPlaceEvent is
 cancelled


diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 015ce2834..e46dd2f59 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -116,12 +116,15 @@ public final class ItemStack {
                 }
             }
         }
+        world.captureDrops = new java.util.ArrayList<>(); // Paper - Don't drop items if block place is cancelled
         boolean flag = this.getItem().interactWith(this, entityhuman, world, blockposition, enumdirection, f, f1, f2);
         int newData = this.getData();
         int newCount = this.count;
         this.count = count;
         this.setData(data);
         world.captureBlockStates = false;
+        java.util.ArrayList<EntityItem> drops = world.captureDrops; // Paper - Don't drop items if block place is cancelled
+        world.captureDrops = null; // Paper - Don't drop items if block place is cancelled
         if (flag && world.captureTreeGeneration && world.capturedBlockStates.size() > 0) {
             world.captureTreeGeneration = false;
             Location location = new Location(world.getWorld(), blockposition.getX(), blockposition.getY(), blockposition.getZ());
@@ -162,10 +165,12 @@ public final class ItemStack {
 
             if (placeEvent != null && (placeEvent.isCancelled() || !placeEvent.canBuild())) {
                 flag = false; // cancel placement
+                world.captureDrops = drops; // Paper - Don't drop items if block place is cancelled
                 // revert back all captured blocks
                 for (BlockState blockstate : blocks) {
                     blockstate.update(true, false);
                 }
+                world.captureDrops = null; // Paper - Don't drop items if block place is cancelled
             } else {
                 // Change the stack to its new contents if it hasn't been tampered with.
                 if (this.count == count && this.getData() == data) {
@@ -188,6 +193,7 @@ public final class ItemStack {
 
                     world.notifyAndUpdatePhysics(newblockposition, null, oldBlock, block.getBlock(), updateFlag); // send null chunk as chunk.k() returns false by this point
                 }
+                if (drops != null) drops.forEach(world::addEntity); // Paper - Don't drop items if block place is cancelled
 
                 for (Map.Entry<BlockPosition, TileEntity> e : world.capturedTileEntities.entrySet()) {
                     world.setTileEntity(e.getKey(), e.getValue());
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 3a6f5299a..ae1fc6005 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -120,6 +120,7 @@ public abstract class World implements IBlockAccess {
             return super.add( blockState );
         }
     };
+    public ArrayList<EntityItem> captureDrops = null; // Paper
     public long ticksPerAnimalSpawns;
     public long ticksPerMonsterSpawns;
     public boolean populating;
@@ -577,7 +578,7 @@ public abstract class World implements IBlockAccess {
             try {
                 // CraftBukkit start
                 CraftWorld world = ((WorldServer) this).getWorld();
-				// TacoSpigot start - Add config to disable redstone firing BlockPhysicsEvent 
+				// TacoSpigot start - Add config to disable redstone firing BlockPhysicsEvent
 				if (world != null && (this.tacoSpigotConfig.isRedstoneFireBPE || !(block instanceof BlockRedstoneWire || block instanceof BlockRedstoneTorch || block instanceof BlockRepeater))) {
 				// TacoSpigot end
                     BlockPhysicsEvent event = new BlockPhysicsEvent(world.getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), CraftMagicNumbers.getId(block));
@@ -691,7 +692,7 @@ public abstract class World implements IBlockAccess {
         }
     }
     // Paper end
-    
+
     public int getLightLevel(BlockPosition blockposition) {
         return this.c(blockposition, true);
     }
@@ -1063,6 +1064,12 @@ public abstract class World implements IBlockAccess {
         org.spigotmc.AsyncCatcher.catchOp( "entity add"); // Spigot
         if (entity == null) return false;
         // CraftBukkit end
+        // Paper start
+        if (captureDrops != null && entity instanceof EntityItem) {
+            captureDrops.add((EntityItem) entity);
+            return false;
+        }
+        // Paper end
         int i = MathHelper.floor(entity.locX / 16.0D);
         int j = MathHelper.floor(entity.locZ / 16.0D);
         boolean flag = entity.attachedToPlayer;
-- 
2.50.1 (Apple Git-155)

