From b9aba28b9d2eda54894ab88d7993b440841801a9 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 29 Nov 2020 20:26:27 +0000
Subject: [PATCH] Allow completely disabling pushing


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 045f0c4a0..6301d5f74 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1703,6 +1703,11 @@ public abstract class EntityLiving extends Entity {
     protected void doTick() {}
 
     protected void bL() {
+        // Blossom start - Allow completely disabling pushing
+        if (world.spigotConfig.maxCollisionsPerEntity == -1) {
+            return;
+        }
+        // Blossom end
         List list = this.world.a((Entity) this, this.getBoundingBox().grow(0.20000000298023224D, 0.0D, 0.20000000298023224D), Predicates.and(IEntitySelector.d, new Predicate() {
             public boolean a(Entity entity) {
                 return entity.ae();
-- 
2.35.2.windows.1

