From 656f8ef96aea9450069b98e7bf7cb1dd0cde97a2 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 27 Nov 2020 22:02:11 +0000
Subject: [PATCH] Merge Spawned Items


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 55ddc7560..f49a46eb3 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -243,4 +243,9 @@ public class BlossomConfig {
         hopperInterval = Math.max(getInt("items.hoppers.delay", 0), 0) + 1;
     }
 
+    public static boolean mergeSpawnedItems;
+    private static void MergeSpawnedItems() {
+        mergeSpawnedItems = getBoolean("items.merged-spawned-items", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index a7281229f..4b16e7218 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -78,6 +78,7 @@ public class Chunk {
     public AtomicInteger pendingLightUpdates = new AtomicInteger();
     public long lightUpdateTime;
     // PaperSpigot end
+    public EntityItem cachedEntityItem; // Blossom - Merge Spawned Items
 
     // PaperSpigot start - ChunkMap caching
     private PacketPlayOutMapChunk.ChunkMap chunkMap;
diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
index 90f58516c..65e2b85e9 100644
--- a/src/main/java/net/minecraft/server/EntityItem.java
+++ b/src/main/java/net/minecraft/server/EntityItem.java
@@ -173,7 +173,7 @@ public class EntityItem extends Entity implements HopperPusher {
 
     }
 
-    private boolean a(EntityItem entityitem) {
+    public boolean a(EntityItem entityitem) { // Blossom - private -> public
         if (entityitem == this) {
             return false;
         } else if (entityitem.isAlive() && this.isAlive()) {
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 565b55612..e53a79047 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1198,7 +1198,29 @@ public abstract class World implements IBlockAccess {
         return true;
     }
 
+    // Blossom start - Merge Spawned Items
+    public boolean addEntity(EntityItem entityItem) {
+        if (me.samsuik.blossom.configuration.BlossomConfig.mergeSpawnedItems) {
+            int floorX = org.bukkit.util.NumberConversions.floor(entityItem.locX) >> 4;
+            int floorZ = org.bukkit.util.NumberConversions.floor(entityItem.locZ) >> 4;
+            Chunk chunk = getChunkIfLoaded(floorX, floorZ);
+
+            EntityItem cachedItem = chunk.cachedEntityItem;
+            if (cachedItem != null && cachedItem.isAlive() &&
+                cachedItem.h(entityItem) < spigotConfig.itemMerge * spigotConfig.itemMerge &&
+                cachedItem.a(entityItem) && !entityItem.isAlive()) {
+                return false;
+            }
+
+            chunk.cachedEntityItem = entityItem;
+        }
+        return addEntity(entityItem, SpawnReason.DEFAULT);
+    }
     public boolean addEntity(Entity entity) {
+        if (entity instanceof EntityItem) {
+            return addEntity((EntityItem) entity);
+        }
+        // Blossom end
         // CraftBukkit start - Used for entities other than creatures
         return addEntity(entity, SpawnReason.DEFAULT);
     }
-- 
2.40.0.windows.1

