From 9599db6de60886d36ffe8cbecaa6cb680bb4a8b7 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 13 Jun 2022 02:09:29 +0100
Subject: [PATCH] Customise Version Command


diff --git a/src/main/java/co/aikar/util/LoadingMap.java b/src/main/java/co/aikar/util/LoadingMap.java
index a9f29199..3d7f53e8 100644
--- a/src/main/java/co/aikar/util/LoadingMap.java
+++ b/src/main/java/co/aikar/util/LoadingMap.java
@@ -27,7 +27,7 @@ package co.aikar.util;
 import com.google.common.base.Function;
 import org.bukkit.Material;
 import co.aikar.timings.TimingHistory;
-import org.w3c.dom.css.Counter;
+//import org.w3c.dom.css.Counter;
 
 import java.lang.reflect.Constructor;
 import java.util.AbstractMap;
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 12d9232c..89a8bd29 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -28,8 +28,10 @@ public class SimpleCommandMap implements CommandMap {
         setDefaultCommands();
     }
 
+    public static VersionCommand versionCommand; // Blossom - I don't like plugins that try to hide what they're using
+
     private void setDefaultCommands() {
-        register("bukkit", new VersionCommand("version"));
+        register("bukkit", versionCommand = new VersionCommand("version")); // Blossom
         register("bukkit", new ReloadCommand("reload"));
         register("bukkit", new PluginsCommand("plugins"));
         register("bukkit", new co.aikar.timings.TimingsCommand("timings")); // Spigot
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 42da60a8..78b0ff60 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -32,6 +32,12 @@ import org.json.simple.parser.ParseException;
 import java.io.InputStreamReader;
 import java.net.HttpURLConnection;
 // TacoSpigot end
+// Blossom start
+import net.md_5.bungee.api.chat.ClickEvent;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+import net.md_5.bungee.api.chat.HoverEvent;
+import net.md_5.bungee.api.chat.TextComponent;
+// Blossom end
 
 public class VersionCommand extends BukkitCommand {
     public VersionCommand(String name) {
@@ -40,17 +46,30 @@ public class VersionCommand extends BukkitCommand {
         this.description = "Gets the version of this server including any plugins in use";
         this.usageMessage = "/version [plugin name]";
         this.setPermission("bukkit.command.version");
-        this.setAliases(Arrays.asList("ver", "about"));
+        this.setAliases(Arrays.asList("ver", "about", "bve"));
     }
 
     @Override
     public boolean execute(CommandSender sender, String currentAlias, String[] args) {
-        if (!testPermission(sender)) return true;
+        // Blossom - moved test permission down
 
         if (args.length == 0) {
-            sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")");
-            sendVersion(sender);
-        } else {
+            // Blossom start - "tidy" this up
+            sender.sendMessage(new ComponentBuilder("")
+                    .append(ChatColor.GRAY + " .\n")
+                    .append(ChatColor.translateAlternateColorCodes('&', " &7| &fThis server is running &eBlos&csom!\n"))
+                    .append(ChatColor.translateAlternateColorCodes('&', " &7| &fCommit&8: &e" + commitHash() + " &7targeting &8(&eMC&8: &71.8.8&8)\n"))
+                    .append(ChatColor.translateAlternateColorCodes('&', " &7| &fGithub&8: &8<"))
+                    .append(ChatColor.translateAlternateColorCodes('&', "&e&olink?"))
+                    .event(new ClickEvent(ClickEvent.Action.OPEN_URL, "https://github.com/Samsuik/Blossom"))
+                    .event(new HoverEvent(HoverEvent.Action.SHOW_TEXT, TextComponent.fromLegacyText(
+                            ChatColor.translateAlternateColorCodes('&', "&o&e*psst!* &fclick here!")
+                    )))
+                    .append(ChatColor.translateAlternateColorCodes('&', "&8>\n"))
+                    .append(ChatColor.GRAY + " '").create());
+            // sendVersion(sender); // Ahhh! this is not good!
+        } else if (testPermission(sender)) {
+            // Blossom end
             StringBuilder name = new StringBuilder();
 
             for (String arg : args) {
@@ -134,7 +153,7 @@ public class VersionCommand extends BukkitCommand {
         Validate.notNull(args, "Arguments cannot be null");
         Validate.notNull(alias, "Alias cannot be null");
 
-        if (args.length == 1) {
+        if (args.length == 1 && testPermission(sender)) { // Blossom
             List<String> completions = new ArrayList<String>();
             String toComplete = args[0].toLowerCase();
             for (Plugin plugin : Bukkit.getPluginManager().getPlugins()) {
@@ -147,6 +166,15 @@ public class VersionCommand extends BukkitCommand {
         return ImmutableList.of();
     }
 
+    // Blossom start
+    public String commitHash() {
+        String string = Bukkit.getVersion();
+        if (string != null && string.startsWith("git-"))
+            string = string.replaceAll(".+?-(.+?) .+$", "$1");
+        return string == null || string.contains("null") ? "in-dev" : string;
+    }
+    // Blossom end
+
     private final ReentrantLock versionLock = new ReentrantLock();
     private boolean hasVersion = false;
     private String versionMessage = null;
-- 
2.40.0.windows.1

