From 175a0f680421c75ecf6eaf5fb177b657eb301942 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 23 Jun 2023 02:00:37 +0100
Subject: [PATCH] Discard entity metadata on removal


diff --git a/src/main/java/org/bukkit/metadata/MetadataStoreBase.java b/src/main/java/org/bukkit/metadata/MetadataStoreBase.java
index 093c1445..c020a6c8 100644
--- a/src/main/java/org/bukkit/metadata/MetadataStoreBase.java
+++ b/src/main/java/org/bukkit/metadata/MetadataStoreBase.java
@@ -8,6 +8,12 @@ import java.util.*;
 public abstract class MetadataStoreBase<T> {
     private Map<String, Map<Plugin, MetadataValue>> metadataMap = new HashMap<String, Map<Plugin, MetadataValue>>();
 
+    // Blossom start
+    public void remove(T subject, String key) {
+        metadataMap.remove(disambiguate(subject, key));
+    }
+    // Blossom end
+
     /**
      * Adds a metadata value to an object. Each metadata value is owned by a
      * specific {@link Plugin}. If a plugin has already added a metadata value
-- 
2.40.0.windows.1

