From f21f038c76a5ae0883a452aebbaf0e267774f3c3 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 16 Jun 2024 11:57:23 +0100
Subject: [PATCH] Modify material size without reflection


diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 0fb26055..f7d66c37 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -461,7 +461,12 @@ public enum Material {
     private final Constructor<? extends MaterialData> ctor;
     private static Material[] byId = new Material[383];
     private final static Map<String, Material> BY_NAME = Maps.newHashMap();
-    private final int maxStack;
+    // Blossom start
+    private int maxStack;
+    public void setMaxStackSize(int size) {
+        this.maxStack = size;
+    }
+    // Blossom end
     private final short durability;
 
     private Material(final int id) {
-- 
2.40.0.windows.1

