From c2c4b532a4eb1c40fabbcf933a19307dae512f17 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 13 Jun 2022 02:49:50 +0100
Subject: [PATCH] Blossom Configuration File


diff --git a/src/main/java/me/samsuik/blossom/commands/BaseCommand.java b/src/main/java/me/samsuik/blossom/commands/BaseCommand.java
new file mode 100644
index 000000000..a52b43d55
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/commands/BaseCommand.java
@@ -0,0 +1,18 @@
+package me.samsuik.blossom.commands;
+
+import org.bukkit.command.Command;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class BaseCommand extends Command {
+    static final Map<Class<?>, BaseCommand> COMMAND_MAP = new HashMap<>();
+    static final Collection<BaseCommand> COMMANDS = COMMAND_MAP.values();
+
+    protected BaseCommand(String name) {
+        super(name);
+        this.setPermission("blossom." + name);
+        BaseCommand.COMMAND_MAP.put(getClass(), this);
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/commands/BlossomCommand.java b/src/main/java/me/samsuik/blossom/commands/BlossomCommand.java
new file mode 100644
index 000000000..f33e04bc9
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/commands/BlossomCommand.java
@@ -0,0 +1,59 @@
+package me.samsuik.blossom.commands;
+
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+import org.bukkit.ChatColor;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.SimpleCommandMap;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public final class BlossomCommand extends Command {
+    public BlossomCommand() {
+        super("blossom");
+    }
+
+    @Override
+    public boolean execute(final CommandSender sender, final String label, final String[] args) {
+        if (args.length > 0) {
+            final List<Command> commands = new ArrayList<>(BaseCommand.COMMANDS);
+            commands.add(SimpleCommandMap.versionCommand);
+
+            for (final Command base : commands) {
+                if (base.getName().equalsIgnoreCase(args[0])) {
+                    return base.execute(sender, label, Arrays.copyOfRange(args, 1, args.length));
+                }
+            }
+        }
+
+        sender.sendMessage(this.createHelpMessage(sender));
+        return false;
+    }
+
+    private BaseComponent[] createHelpMessage(final CommandSender sender) {
+        final ComponentBuilder builder = new ComponentBuilder("")
+                .append(ChatColor.GRAY + " .\n")
+                .append(ChatColor.translateAlternateColorCodes('&', "&7 | &fThis is the main command for &eBlos&csom.\n"))
+                .append(ChatColor.translateAlternateColorCodes('&', "&7 | &fAll exclusive commands are listed below.\n"))
+                .append(ChatColor.GRAY + " |\n");
+
+        for (final BaseCommand base : BaseCommand.COMMANDS) {
+            if (!base.testPermissionSilent(sender)) {
+                continue;
+            }
+
+            builder.append(ChatColor.translateAlternateColorCodes('&', " &7| &8* &7/&e") + base.getName());
+
+            if (!base.getDescription().equals("")) {
+                builder.append(ChatColor.translateAlternateColorCodes('&', "&8; &f") + base.getDescription());
+            }
+
+            builder.append("\n");
+        }
+
+        return builder.append(ChatColor.GRAY + " '").create();
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/commands/BlossomCommands.java b/src/main/java/me/samsuik/blossom/commands/BlossomCommands.java
new file mode 100644
index 000000000..424b9c371
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/commands/BlossomCommands.java
@@ -0,0 +1,13 @@
+package me.samsuik.blossom.commands;
+
+import me.samsuik.blossom.commands.subcommands.ConfigCommand;
+import org.bukkit.command.Command;
+
+import java.util.Map;
+
+public class BlossomCommands {
+    public static void register(final Map<String, Command> commands) {
+        commands.put("blossom", new BlossomCommand());
+        commands.put("config", new ConfigCommand());
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/commands/subcommands/ConfigCommand.java b/src/main/java/me/samsuik/blossom/commands/subcommands/ConfigCommand.java
new file mode 100644
index 000000000..9456e5d03
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/commands/subcommands/ConfigCommand.java
@@ -0,0 +1,37 @@
+package me.samsuik.blossom.commands.subcommands;
+
+import me.samsuik.blossom.commands.BaseCommand;
+import net.md_5.bungee.api.ChatColor;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.World;
+import org.bukkit.command.CommandSender;
+
+import java.io.File;
+
+public final class ConfigCommand extends BaseCommand {
+    public ConfigCommand() {
+        super("config");
+        this.setPermission("blossom.config");
+    }
+
+    @Override
+    public boolean execute(final CommandSender sender, final String label, final String[] args) {
+        if (!this.testPermission(sender)) {
+            return true;
+        }
+
+        me.samsuik.blossom.configuration.BlossomConfig.init((File) MinecraftServer.getServer().options.valueOf("blossom-settings"));
+        for (final World world : MinecraftServer.getServer().worlds) {
+            world.blossomConfig.init();
+        }
+        me.samsuik.blossom.configuration.BlossomConfig.registerCommands();
+
+        sender.sendMessage(new ComponentBuilder("")
+                .append(org.bukkit.ChatColor.GRAY + " .\n")
+                .append(ChatColor.translateAlternateColorCodes('&', " &7| &fSuccessfully reloaded the &eBlos&csom&f configuration file.\n"))
+                .append(org.bukkit.ChatColor.GRAY + " '").create());
+
+        return true;
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
new file mode 100644
index 000000000..f283caf6d
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -0,0 +1,128 @@
+package me.samsuik.blossom.configuration;
+
+import com.google.common.base.Throwables;
+import me.samsuik.blossom.commands.BlossomCommands;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.command.Command;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public final class BlossomConfig {
+
+    private static final Logger LOGGER = Logger.getLogger("Blossom");
+
+    private static File CONFIG_FILE;
+    private static final String HEADER = "This is the main configuration file for Blossom.\n";
+    /*========================================================================*/
+    public static YamlConfiguration config;
+    static int version;
+    static Map<String, Command> commands;
+    /*========================================================================*/
+
+    public static void init(final File file) {
+        CONFIG_FILE = file;
+        config = new YamlConfiguration();
+
+        try {
+            config.load(CONFIG_FILE);
+        } catch (final IOException ignored) {
+        } catch (final InvalidConfigurationException ex) {
+            LOGGER.log(Level.SEVERE, "Could not load blossom.yml, please correct your syntax errors", ex);
+            throw Throwables.propagate(ex);
+        }
+
+        config.options().header(HEADER);
+        config.options().copyDefaults(true);
+
+        commands = new HashMap<>();
+        BlossomCommands.register(commands);
+
+        version = getInt("config-version", 4);
+        set("config-version", 4);
+        readConfig(BlossomConfig.class, null);
+    }
+
+    public static void registerCommands() {
+        for (final Map.Entry<String, Command> entry : commands.entrySet()) {
+            MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Blossom", entry.getValue());
+        }
+    }
+
+    static void readConfig(final Class<?> clazz, final Object instance) {
+        for (final Method method : clazz.getDeclaredMethods()) {
+            if (Modifier.isPrivate(method.getModifiers())) {
+                if (method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE) {
+                    try {
+                        method.setAccessible(true);
+                        method.invoke(instance);
+                    } catch (InvocationTargetException ex) {
+                        throw Throwables.propagate(ex.getCause());
+                    } catch (Exception ex) {
+                        LOGGER.log(Level.SEVERE, "Error invoking " + method, ex);
+                    }
+                }
+            }
+        }
+
+        try {
+            config.save(CONFIG_FILE);
+        } catch (final IOException ex) {
+            LOGGER.log(Level.SEVERE, "Could not save " + CONFIG_FILE, ex);
+        }
+    }
+
+    private static void set(final String path, final Object val) {
+        config.set(path, val);
+    }
+
+    private static void remove(final String path) {
+        set(path, null);
+    }
+
+    private static boolean getBoolean(final String path, final boolean def) {
+        config.addDefault(path, def);
+        return config.getBoolean(path, config.getBoolean(path));
+    }
+
+    private static double getDouble(final String path, final double def) {
+        config.addDefault(path, def);
+        return config.getDouble(path, config.getDouble(path));
+    }
+
+    private static long getLong(final String path, final long def) {
+        config.addDefault(path, def);
+        return config.getLong(path, config.getLong(path));
+    }
+
+    private static int getInt(final String path, final int def) {
+        config.addDefault(path, def);
+        return config.getInt(path, config.getInt(path));
+    }
+
+    private static <T> List<T> getList(final String path, final T def) {
+        config.addDefault(path, def);
+        return (List<T>) config.getList(path, config.getList(path));
+    }
+
+    private static String getString(final String path, final String def) {
+        config.addDefault(path, def);
+        return config.getString(path, config.getString(path));
+    }
+
+    private static void remove() {
+        remove("command-signs");
+        remove("books.character-checking");
+    }
+
+}
diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomWorldConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomWorldConfig.java
new file mode 100644
index 000000000..98b731347
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomWorldConfig.java
@@ -0,0 +1,63 @@
+package me.samsuik.blossom.configuration;
+
+import java.util.List;
+
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+public final class BlossomWorldConfig {
+
+    private final String worldName;
+    private final YamlConfiguration config;
+
+    public BlossomWorldConfig(final String worldName) {
+        this.worldName = worldName;
+        this.config = BlossomConfig.config;
+        this.init();
+    }
+
+    public void init() {
+        BlossomConfig.readConfig(BlossomWorldConfig.class, this);
+    }
+
+    private void set(final String path, final Object val) {
+        this.config.set(path, val);
+    }
+
+    private void remove(final String path) {
+        set(path, null);
+    }
+
+    private String path(final String path) {
+        return "world-settings." + this.worldName + "." + path;
+    }
+
+    private boolean getBoolean(final String path, final boolean def) {
+        this.config.addDefault(path, def);
+        return this.config.getBoolean(this.path(path), this.config.getBoolean(path));
+    }
+
+    private double getDouble(final String path, final double def) {
+        this.config.addDefault(path, def);
+        return this.config.getDouble(this.path(path), this.config.getDouble(path));
+    }
+
+    private int getInt(final String path, final int def) {
+        this.config.addDefault(path, def);
+        return this.config.getInt(this.path(path), this.config.getInt(path));
+    }
+
+    private <T extends List<?>> T getList(final String path, final T def) {
+        this.config.addDefault(path, def);
+        return (T) this.config.getList(this.path(path), this.config.getList(path));
+    }
+
+    private String getString(final String path, final String def) {
+        this.config.addDefault(path, def);
+        return this.config.getString(this.path(path), this.config.getString(path));
+    }
+
+    private void remove() {
+    }
+
+}
diff --git a/src/main/java/me/samsuik/blossom/util/MessageUtil.java b/src/main/java/me/samsuik/blossom/util/MessageUtil.java
new file mode 100644
index 000000000..0f3b1dc03
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/util/MessageUtil.java
@@ -0,0 +1,39 @@
+package me.samsuik.blossom.util;
+
+import org.bukkit.ChatColor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class MessageUtil {
+    private static final Pattern TAG_PATTERN = Pattern.compile("%([^%]+)?%?");
+
+    public static String format(String message, final Object... args) {
+        final Map<String, String> placeholders = new HashMap<>();
+        for (int index = 0; index < args.length; index += 2) {
+            placeholders.put(args[index].toString(), args[index + 1].toString());
+        }
+
+        final Matcher tagMatcher = TAG_PATTERN.matcher(message);
+        while (tagMatcher.find()) {
+            final String tag = tagMatcher.group(1);
+            final String[] parts = tag.split("\\$");
+
+            String placeholder = placeholders.get(parts[0]);
+            for (int i = 1; i < parts.length; i++) {
+                final String part = parts[i];
+                final String[] replacement = part.split("=");
+
+                if (replacement.length == 2) {
+                    placeholder = placeholder.replaceAll(replacement[0], replacement[1]);
+                }
+            }
+
+            message = message.replace("%" + tag + "%", placeholder);
+        }
+
+        return ChatColor.translateAlternateColorCodes('&', message);
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/util/ModernMaterial.java b/src/main/java/me/samsuik/blossom/util/ModernMaterial.java
new file mode 100644
index 000000000..43ad0a419
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/util/ModernMaterial.java
@@ -0,0 +1,91 @@
+package me.samsuik.blossom.util;
+
+import org.bukkit.Material;
+import org.bukkit.inventory.ItemStack;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class ModernMaterial {
+    private final Material material;
+    private final int data;
+
+    public static int getData(final String materialName) {
+        return getMaterial(materialName).getData();
+    }
+
+    public static ModernMaterial getMaterial(final String materialName) {
+        final Matcher extracted = DataMapping.DATA_PATTERN.matcher(materialName);
+        if (extracted.find()) {
+            final int data = DataMapping.MATERIAL_DATA.getOrDefault(extracted.group(1), 0);
+            final String legacyMaterial = extracted.group(2);
+            return new ModernMaterial(Material.getMaterial(legacyMaterial), data);
+        } else {
+            return new ModernMaterial(Material.getMaterial(materialName), 0);
+        }
+    }
+
+    public ModernMaterial(final Material material, final int data) {
+        this.material = material;
+        this.data = data;
+    }
+
+    public Material getMaterial() {
+        return this.material;
+    }
+
+    public int getData() {
+        return this.data;
+    }
+
+    public ItemStack createItem(final int count) {
+        return new ItemStack(this.material, count, (short) this.data);
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (!(o instanceof ModernMaterial)) return false;
+        final ModernMaterial that = (ModernMaterial) o;
+        return data == that.data && material == that.material;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(material, data);
+    }
+
+    @Override
+    public String toString() {
+        return "ModernMaterial{" +
+                "material=" + material +
+                ", data=" + data +
+                '}';
+    }
+
+    private static final class DataMapping {
+        private static final List<String> COLORS;
+        private static final Map<String, Integer> MATERIAL_DATA = new HashMap<>();
+        private static final Pattern DATA_PATTERN;
+
+        static {
+            COLORS = Arrays.asList(
+                "WHITE", "ORANGE", "MAGENTA", "LIGHT_BLUE", "YELLOW", "LIME", "PINK", "GRAY", "LIGHT_GRAY", "CYAN", "PURPLE", "BLUE", "BROWN", "GREEN", "RED", "BLACK"
+            );
+
+            for (int index = 0; index < COLORS.size(); index++) {
+                MATERIAL_DATA.put(COLORS.get(index), index);
+            }
+
+            final StringBuilder builder = new StringBuilder();
+            for (final String color : COLORS) {
+                if (builder.length() > 0) {
+                    builder.append('|');
+                }
+                builder.append(color);
+            }
+
+            DATA_PATTERN = Pattern.compile("^(" + builder.toString() + ")_(.+)");
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index 088beb22b..d2013851b 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -181,6 +181,10 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
             org.github.paperspigot.PaperSpigotConfig.init((File) options.valueOf("paper-settings"));
             org.github.paperspigot.PaperSpigotConfig.registerCommands();
             // PaperSpigot end
+            // Blossom start - blossom configuration files
+            me.samsuik.blossom.configuration.BlossomConfig.init((File) options.valueOf("blossom-settings"));
+            me.samsuik.blossom.configuration.BlossomConfig.registerCommands();
+            // Blossom end - blossom configuration files
 
             DedicatedServer.LOGGER.info("Generating keypair");
             this.a(MinecraftEncryption.b());
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 8780cca2f..a0a445f18 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -164,6 +164,7 @@ public abstract class World implements IBlockAccess {
 
     public final co.aikar.timings.WorldTimingsHandler timings; // Spigot
     public final net.techcable.tacospigot.TacoSpigotWorldConfig tacoSpigotConfig;
+    public final me.samsuik.blossom.configuration.BlossomWorldConfig blossomConfig; // Blossom - blossom configuration files
 
     public CraftWorld getWorld() {
         return this.world;
@@ -187,6 +188,7 @@ public abstract class World implements IBlockAccess {
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig( worlddata.getName() ); // Spigot
         this.paperSpigotConfig = new org.github.paperspigot.PaperSpigotWorldConfig( worlddata.getName() ); // PaperSpigot
         this.tacoSpigotConfig = new net.techcable.tacospigot.TacoSpigotWorldConfig(worlddata.getName()); // TacoSpigot
+        this.blossomConfig = new me.samsuik.blossom.configuration.BlossomWorldConfig(worlddata.getName()); // Blossom - blossom configuration files
         this.generator = gen;
         this.world = new CraftWorld((WorldServer) this, gen, env);
         this.ticksPerAnimalSpawns = this.getServer().getTicksPerAnimalSpawns(); // CraftBukkit
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 17831f9f4..9bbb6c43a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -715,6 +715,7 @@ public final class CraftServer implements Server {
         org.spigotmc.SpigotConfig.init((File) console.options.valueOf("spigot-settings")); // Spigot
         org.github.paperspigot.PaperSpigotConfig.init((File) console.options.valueOf("paper-settings")); // PaperSpigot
         net.techcable.tacospigot.TacoSpigotConfig.init((File) console.options.valueOf("taco-settings")); // TacoSpigot
+        me.samsuik.blossom.configuration.BlossomConfig.init((File) console.options.valueOf("blossom-settings")); // Blossom - blossom configuration files
         for (WorldServer world : console.worlds) {
             world.worldData.setDifficulty(difficulty);
             world.setSpawnFlags(monsters, animals);
@@ -732,6 +733,7 @@ public final class CraftServer implements Server {
             world.spigotConfig.init(); // Spigot
             world.paperSpigotConfig.init(); // PaperSpigot
             world.tacoSpigotConfig.init(); // TacoSpigot
+            world.blossomConfig.init(); // Blossom - blossom configuration files
         }
 
         pluginManager.clearPlugins();
@@ -739,6 +741,7 @@ public final class CraftServer implements Server {
         resetRecipes();
         org.spigotmc.SpigotConfig.registerCommands(); // Spigot
         org.github.paperspigot.PaperSpigotConfig.registerCommands(); // PaperSpigot
+        me.samsuik.blossom.configuration.BlossomConfig.registerCommands(); // Blossom - blossom configuration files
 
         overrideAllCommandBlockCommands = commandsConfiguration.getStringList("command-block-overrides").contains("*");
 
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 569a8846b..210d8b1d3 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -152,6 +152,14 @@ public class Main {
                         .defaultsTo(new File("taco.yml"))
                         .describedAs("Yml file");
                 // TacoSpigot end
+
+                // Blossom start - blossom configuration files
+                acceptsAll(asList("blossom", "blossom-settings"), "File for blossom settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("blossom.yml"))
+                        .describedAs("Yml file");
+                // Blossom end - blossom configuration files
             }
         };
 
@@ -220,6 +228,7 @@ public class Main {
                 }
                 // Spigot End
                 net.techcable.tacospigot.TacoSpigotConfig.init((File) options.valueOf("taco-settings")); // TacoSpigot - load config before we load libraries to allow access while loading
+                me.samsuik.blossom.configuration.BlossomConfig.init((File) options.valueOf("blossom-settings")); // Blossom - blossom configuration files
                 System.out.println("Loading libraries, please wait...");
                 MinecraftServer.main(options);
             } catch (Throwable t) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ad07563d2..788216d03 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -20,6 +20,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+
+import me.samsuik.blossom.visuals.Visuals;
 import net.md_5.bungee.api.chat.BaseComponent;
 
 import net.minecraft.server.*;
-- 
2.50.1 (Apple Git-155)

