From a08b51ede4d848bda0faae7d0575739ae002efad Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 23 May 2023 23:07:20 +0100
Subject: [PATCH] utils


diff --git a/src/main/java/me/samsuik/blossom/util/collections/OrderedComparatorList.java b/src/main/java/me/samsuik/blossom/util/collections/OrderedComparatorList.java
new file mode 100644
index 000000000..2597affed
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/util/collections/OrderedComparatorList.java
@@ -0,0 +1,49 @@
+package me.samsuik.blossom.util.collections;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+
+import java.util.Arrays;
+import java.util.Comparator;
+
+public final class OrderedComparatorList<T> extends ObjectArrayList<T> {
+    private final Comparator<T> comparator;
+    private boolean binarySearch = true;
+
+    public OrderedComparatorList(final int capacity, final Comparator<T> comparator) {
+        super(capacity);
+        this.comparator = Comparator.nullsLast(comparator);
+    }
+
+    public OrderedComparatorList(final Comparator<T> comparator) {
+        this(DEFAULT_INITIAL_CAPACITY, comparator);
+    }
+
+    private void validateBounds(final int index, final T element, final boolean up) {
+        if (index != 0 && this.comparator.compare(get(index - 1), element) > 0) {
+            this.binarySearch = false;
+        } else if (up && index < size() - 1 && this.comparator.compare(get(index + 1), element) < 0) {
+            this.binarySearch = false;
+        }
+    }
+
+    @Override
+    public boolean add(final T element) {
+        this.validateBounds(size(), element, false);
+        return super.add(element);
+    }
+
+    @Override
+    public void add(final int index, final T element) {
+        this.validateBounds(index, element, true);
+        super.add(index, element);
+    }
+
+    @Override
+    public int indexOf(final Object element) {
+        if (this.binarySearch) {
+            return Math.max(Arrays.binarySearch(this.a, (T) element, this.comparator), -1);
+        } else {
+            return super.indexOf(element);
+        }
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/util/collections/UnorderedIndexedList.java b/src/main/java/me/samsuik/blossom/util/collections/UnorderedIndexedList.java
new file mode 100644
index 000000000..4ac236ca8
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/util/collections/UnorderedIndexedList.java
@@ -0,0 +1,62 @@
+package me.samsuik.blossom.util.collections;
+
+import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+
+public final class UnorderedIndexedList<T> extends ObjectArrayList<T> {
+    private final Int2IntOpenHashMap elementToIndex = new Int2IntOpenHashMap();
+    {
+        this.elementToIndex.defaultReturnValue(-1);
+    }
+
+    public UnorderedIndexedList(final int capacity) {
+        super(capacity);
+    }
+
+    public UnorderedIndexedList() {
+        super();
+    }
+
+    @Override
+    public boolean add(final T element) {
+        this.elementToIndex.put(element.hashCode(), this.size());
+        return super.add(element);
+    }
+
+    @Override
+    public T remove(final int index) {
+        final int tail = this.size() - 1;
+        final T at = this.a[index];
+        if (index != tail) {
+            final T tailObj = this.a[tail];
+            if (tailObj != null) {
+                this.elementToIndex.put(tailObj.hashCode(), index);
+            }
+            this.a[index] = tailObj;
+        }
+
+        if (at != null) {
+            this.elementToIndex.remove(at.hashCode());
+        }
+
+        this.a[tail] = null;
+        this.size--;
+        return at;
+    }
+
+    @Override
+    public void clear() {
+        this.elementToIndex.clear();
+        super.clear();
+    }
+
+    @Override
+    public int indexOf(final Object k) {
+        return this.elementToIndex.get(k.hashCode());
+    }
+
+    @Override
+    public void add(final int index, final T element) {
+        throw new UnsupportedOperationException();
+    }
+}
diff --git a/src/main/java/net/minecraft/server/BlockPosition.java b/src/main/java/net/minecraft/server/BlockPosition.java
index ac43e5a90..1ce794bd2 100644
--- a/src/main/java/net/minecraft/server/BlockPosition.java
+++ b/src/main/java/net/minecraft/server/BlockPosition.java
@@ -152,6 +152,32 @@ public class BlockPosition extends BaseBlockPosition {
         return ((long) this.getX() & BlockPosition.h) << BlockPosition.g | ((long) this.getY() & BlockPosition.i) << BlockPosition.f | ((long) this.getZ() & BlockPosition.j) << 0;
     }
 
+    // Blossom start - utils
+    public static long asLong(final int x, final int y, final int z) {
+        return ((long) x & BlockPosition.h) << BlockPosition.g | ((long) y & BlockPosition.i) << BlockPosition.f | ((long) z & BlockPosition.j);
+    }
+
+    public static int getX(final long pack) {
+        return (int) (pack >> BlockPosition.g & BlockPosition.h);
+    }
+
+    public static int getY(final long pack) {
+        return (int) (pack >> BlockPosition.f & BlockPosition.i);
+    }
+
+    public static int getZ(final long pack) {
+        return (int) (pack & BlockPosition.j);
+    }
+
+    public final BlockPosition.MutableBlockPosition mutable() {
+        return new MutableBlockPosition(this.getX(), this.getY(), this.getZ());
+    }
+
+    public final BlockPosition immutable() {
+        return new BlockPosition(this.getX(), this.getY(), this.getZ());
+    }
+    // Blossom end - utils
+
     public static BlockPosition fromLong(long i) {
         int j = (int) (i << 64 - BlockPosition.g - BlockPosition.c >> 64 - BlockPosition.c);
         int k = (int) (i << 64 - BlockPosition.f - BlockPosition.e >> 64 - BlockPosition.e);
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index a7281229f..fba785e19 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -519,6 +519,26 @@ public class Chunk {
     }
 
     // PaperSpigot start - Optimize getBlockData
+    // Blossom start - utils
+    public final boolean isWithin(final int x, final int z) {
+        return this.locX == x && this.locZ == z && this.isLoaded();
+    }
+
+    public final boolean isWithin(final BlockPosition blockposition) {
+        return this.isWithin(blockposition.getX() >> 4, blockposition.getZ() >> 4);
+    }
+
+    public final IBlockData getBlockData(final int x, final int y, final int z) {
+        if (y >= 0 && y >> 4 < this.sections.length) {
+            final ChunkSection chunksection = this.sections[y >> 4];
+            if (chunksection != null) {
+                return chunksection.getType(x & 15, y & 15, z & 15);
+            }
+        }
+        return Blocks.AIR.getBlockData();
+    }
+    // Blossom end - utils
+
     public IBlockData getBlockData(final BlockPosition blockposition) {
         if (blockposition.getY() >= 0 && blockposition.getY() >> 4 < this.sections.length) {
             ChunkSection chunksection = this.sections[blockposition.getY() >> 4];
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index a0a445f18..bc3b18608 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -831,6 +831,12 @@ public abstract class World implements IBlockAccess {
         return null;
     }
     // Paper end
+    // Blossom start - utils
+    public final IBlockData getTypeIfLoaded(final int x, final int y, final int z) {
+        final Chunk chunk = this.getChunkIfLoaded(x >> 4, z >> 4);
+        return chunk != null ? chunk.getBlockData(x, y, z) : null;
+    }
+    // Blossom end - utils
 
     // Spigot start
     public IBlockData getType(BlockPosition blockposition)
-- 
2.50.1 (Apple Git-155)

