From ebc0c1c19c657520e64873ddd9d43f90c76d93a0 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 1 Jul 2022 00:09:43 +0100
Subject: [PATCH] Prevent locking each packet


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 8a88a9aaf..6b91e25a7 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -258,7 +258,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
     // Tuinity end - add flush parameter
 
     private void m() {
-        if (this.channel != null && this.channel.isOpen()) {
+        if (this.channel != null && this.channel.isOpen() && !this.i.isEmpty()) { // Blossom - don't lock here
             this.j.readLock().lock();
 
             // Tuinity start - make only one flush call per sendPacketQueue() call
-- 
2.40.0.windows.1

