From 00fc0ea69919f3cc294e803753889c331699d0af Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 9 May 2023 20:37:38 +0100
Subject: [PATCH] Allow disabling command signs


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 235b49e3c..3eec2ce55 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -194,5 +194,10 @@ public class BlossomConfig {
         bookLength = getInt("books.page-length", 400); // later versions require this to be around 800
         bookPages = getInt("books.max-pages", 50); // later versions can support up to 100 pages, ouch
     }
+
+    public static boolean commandSigns;
+    private static void CommandSign() {
+        commandSigns = getBoolean("command-signs", true);
+    }
     
 }
diff --git a/src/main/java/net/minecraft/server/TileEntitySign.java b/src/main/java/net/minecraft/server/TileEntitySign.java
index e927fd237..5f787d6dc 100644
--- a/src/main/java/net/minecraft/server/TileEntitySign.java
+++ b/src/main/java/net/minecraft/server/TileEntitySign.java
@@ -175,7 +175,7 @@ public class TileEntitySign extends TileEntity {
             if (chatmodifier != null && chatmodifier.h() != null) {
                 ChatClickable chatclickable = chatmodifier.h();
 
-                if (chatclickable.a() == ChatClickable.EnumClickAction.RUN_COMMAND) {
+                if (chatclickable.a() == ChatClickable.EnumClickAction.RUN_COMMAND && me.samsuik.blossom.configuration.BlossomConfig.commandSigns) {
                     // CraftBukkit start
                     // MinecraftServer.getServer().getCommandHandler().a(tileentitysignplayerwrapper, chatclickable.b());
                     CommandBlockListenerAbstract.executeCommand(entityhuman, (org.bukkit.entity.Player) entityhuman.getBukkitEntity(), chatclickable.b());
-- 
2.50.1

