From 7e1b336a8f252d602263ed319117b795ffd2ec5e Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 9 May 2023 21:05:18 +0100
Subject: [PATCH] Restrict anvil item name reading to 30 characters


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 570fd76c4..5af2c9806 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2178,7 +2178,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             ContainerAnvil containeranvil = (ContainerAnvil) this.player.activeContainer;
 
             if (packetplayincustompayload.b() != null && packetplayincustompayload.b().readableBytes() >= 1) {
-                String s1 = SharedConstants.a(packetplayincustompayload.b().c(32767));
+                String s1 = SharedConstants.a(packetplayincustompayload.b().c(30)); // Blossom - do not read strings up to 128kb for item names
 
                 if (s1.length() <= 30) {
                     containeranvil.a(s1);
-- 
2.50.1

