From 7d3e575822b00852701534eadb6787387648a29b Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 5 Jul 2022 18:08:16 +0100
Subject: [PATCH] FPS Command


diff --git a/src/main/java/me/samsuik/blossom/commands/subcommands/FPSCommand.java b/src/main/java/me/samsuik/blossom/commands/subcommands/FPSCommand.java
new file mode 100644
index 000000000..4eac6e8c8
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/commands/subcommands/FPSCommand.java
@@ -0,0 +1,31 @@
+package me.samsuik.blossom.commands.subcommands;
+
+import me.samsuik.blossom.commands.BaseCommand;
+import me.samsuik.blossom.visuals.Visuals;
+import me.samsuik.blossom.visuals.gui.FPSGUI;
+import net.md_5.bungee.api.ChatColor;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+import net.minecraft.server.EntityPlayer;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.entity.Player;
+
+public class FPSCommand extends BaseCommand {
+
+    public FPSCommand() {
+        super("fps");
+    }
+
+    @Override
+    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
+        if (!(sender instanceof Player)) {
+            return true;
+        } else if (!testPermission(sender)) {
+            return true;
+        }
+
+        FPSGUI.getInstance().showTo((Player) sender);
+        return true;
+    }
+
+}
diff --git a/src/main/java/me/samsuik/blossom/commands/subcommands/VisualCommand.java b/src/main/java/me/samsuik/blossom/commands/subcommands/VisualCommand.java
new file mode 100644
index 000000000..410f91574
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/commands/subcommands/VisualCommand.java
@@ -0,0 +1,45 @@
+package me.samsuik.blossom.commands.subcommands;
+
+import me.samsuik.blossom.commands.BaseCommand;
+import me.samsuik.blossom.visuals.Visuals;
+import net.md_5.bungee.api.ChatColor;
+import net.md_5.bungee.api.chat.ComponentBuilder;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Player;
+
+import java.util.Arrays;
+import java.util.Locale;
+
+public class VisualCommand extends BaseCommand {
+
+    private final Visuals.VisualType type;
+
+    public VisualCommand(Visuals.VisualType type, String... aliases) {
+        super(type.name().toLowerCase(Locale.ROOT).replace("_", ""));
+        this.setAliases(Arrays.asList(aliases));
+        this.type = type;
+    }
+
+    @Override
+    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
+        if (!(sender instanceof Player)) {
+            return true;
+        } else if (!testPermission(sender)) {
+            return true;
+        }
+
+        Player player = (Player) sender;
+        Visuals viusals = player.getVisuals();
+
+        viusals.toggle(type);
+        String state = viusals.isEnabled(type) ? "Enabled" : "Disabled";
+
+        sender.sendMessage(new ComponentBuilder("")
+                .append(org.bukkit.ChatColor.GRAY + " .\n")
+                .append(ChatColor.translateAlternateColorCodes('&', " &7| &e" + state + " &7" + type.getFriendlyName() + "\n"))
+                .append(org.bukkit.ChatColor.GRAY + " '").create());
+
+        return true;
+    }
+
+}
diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 3eec2ce55..d2cae0493 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -3,6 +3,9 @@ package me.samsuik.blossom.configuration;
 import com.google.common.base.Throwables;
 import me.samsuik.blossom.commands.BlossomCommand;
 import me.samsuik.blossom.commands.subcommands.ConfigCommand;
+import me.samsuik.blossom.commands.subcommands.FPSCommand;
+import me.samsuik.blossom.commands.subcommands.VisualCommand;
+import me.samsuik.blossom.visuals.Visuals;
 import net.minecraft.server.MinecraftServer;
 import org.bukkit.command.Command;
 import org.bukkit.configuration.InvalidConfigurationException;
@@ -51,6 +54,10 @@ public class BlossomConfig {
 
         commands.put("blossom", new BlossomCommand());
         commands.put("config", new ConfigCommand());
+        commands.put("fps", new FPSCommand());
+        commands.put("tntvisibility", new VisualCommand(Visuals.VisualType.TNT_VISIBILITY, "tnttoggle"));
+        commands.put("sandvisibility", new VisualCommand(Visuals.VisualType.SAND_VISIBILITY, "sandtoggle"));
+        commands.put("minimal", new VisualCommand(Visuals.VisualType.MINIMAL, "minimaltnt", "tntlag"));
 
         version = getInt("config-version", 3);
         set("config-version", 3);
-- 
2.50.1

