From 8893e661ae2d24706ae8172c754c8ecc3239aa07 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 28 Jun 2022 23:06:32 +0100
Subject: [PATCH] Optimise scheduled block ticking and unloading


diff --git a/src/main/java/me/samsuik/blossom/world/ScheduledBlock.java b/src/main/java/me/samsuik/blossom/world/ScheduledBlock.java
new file mode 100644
index 000000000..28caf35b3
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/world/ScheduledBlock.java
@@ -0,0 +1,37 @@
+package me.samsuik.blossom.world;
+
+import it.unimi.dsi.fastutil.HashCommon;
+import net.minecraft.server.Block;
+import net.minecraft.server.BlockPosition;
+import net.minecraft.server.NextTickListEntry;
+
+import java.util.Objects;
+
+public final class ScheduledBlock {
+    private final Block type;
+    private final BlockPosition pos;
+
+    public static ScheduledBlock from(final NextTickListEntry entry) {
+        return new ScheduledBlock(entry.e, entry.a);
+    }
+
+    public ScheduledBlock(final Block type, final BlockPosition pos) {
+        this.type = type;
+        this.pos = pos;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof ScheduledBlock)) return false;
+
+        final ScheduledBlock that = (ScheduledBlock) o;
+        return Objects.equals(this.type, that.type) && Objects.equals(this.pos, that.pos);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = HashCommon.mix(this.type.hashCode());
+        result = 31 * result + Long.hashCode(pos.asLong());
+        return result;
+    }
+}
diff --git a/src/main/java/me/samsuik/blossom/world/ScheduledBlockTicks.java b/src/main/java/me/samsuik/blossom/world/ScheduledBlockTicks.java
new file mode 100644
index 000000000..989cf6752
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/world/ScheduledBlockTicks.java
@@ -0,0 +1,73 @@
+package me.samsuik.blossom.world;
+
+import net.minecraft.server.NextTickListEntry;
+
+import java.util.*;
+
+public final class ScheduledBlockTicks extends AbstractSet<NextTickListEntry> {
+    private final HashSet<ScheduledBlock> scheduledBlocks = new HashSet<>();
+    private final TreeSet<NextTickListEntry> blocksToTick = new TreeSet<>();
+
+    @Override
+    public int size() {
+        return this.scheduledBlocks.size();
+    }
+
+    @Override
+    public boolean add(final NextTickListEntry entry) {
+        final ScheduledBlock scheduledBlock = new ScheduledBlock(entry.e, entry.a);
+        if (this.scheduledBlocks.add(scheduledBlock)) {
+            this.blocksToTick.add(entry);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean remove(final Object o) {
+        if (!(o instanceof NextTickListEntry)) return false;
+        final NextTickListEntry entry = (NextTickListEntry) o;
+
+        this.blocksToTick.remove(entry);
+        return this.scheduledBlocks.remove(new ScheduledBlock(entry.e, entry.a));
+    }
+
+    @Override
+    public void clear() {
+        this.scheduledBlocks.clear();
+        this.blocksToTick.clear();
+    }
+
+    @Override
+    public Iterator<NextTickListEntry> iterator() {
+        return new Iterator<NextTickListEntry>() {
+            private final Iterator<NextTickListEntry> it = ScheduledBlockTicks.this.blocksToTick.iterator();
+            private NextTickListEntry last;
+
+            @Override
+            public boolean hasNext() {
+                return this.it.hasNext();
+            }
+
+            @Override
+            public NextTickListEntry next() {
+                return this.last = it.next();
+            }
+
+            @Override
+            public void remove() {
+                final NextTickListEntry entry = this.last;
+                if (entry == null) {
+                    throw new IllegalStateException("No element to remove");
+                }
+
+                this.last = null;
+                this.it.remove();
+
+                final ScheduledBlock scheduledBlock = new ScheduledBlock(entry.e, entry.a);
+                ScheduledBlockTicks.this.scheduledBlocks.remove(scheduledBlock);
+            }
+        };
+    }
+}
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index fe82dc3db..7cfe645cb 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -75,7 +75,8 @@ public class Chunk {
     private ConcurrentLinkedQueue<BlockPosition> w;
     protected gnu.trove.map.hash.TObjectIntHashMap<Class> entityCount = new gnu.trove.map.hash.TObjectIntHashMap<Class>(); // Spigot
     protected it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<Class<?>, List<Entity>>[] classToEntityListMap = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap[16]; // Blossom
-    public List<Entity>[] hardColliding = new me.samsuik.blossom.utils.collections.UnorderedIndexedList[16]; // Blossom
+    public List<Entity>[] hardColliding = new me.samsuik.blossom.util.collections.UnorderedIndexedList[16]; // Blossom
+    public final java.util.Set<NextTickListEntry> scheduledToTick = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(); // Blossom - optimise scheduled block ticking
     // PaperSpigot start - Asynchronous light updates
     public AtomicInteger pendingLightUpdates = new AtomicInteger();
     public long lightUpdateTime;
diff --git a/src/main/java/net/minecraft/server/ChunkRegionLoader.java b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
index e867a944f..3eb29e754 100644
--- a/src/main/java/net/minecraft/server/ChunkRegionLoader.java
+++ b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
@@ -464,7 +464,7 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
                         block = Block.getById(nbttagcompound5.getInt("i"));
                     }
 
-                    world.b(new BlockPosition(nbttagcompound5.getInt("x"), nbttagcompound5.getInt("y"), nbttagcompound5.getInt("z")), block, nbttagcompound5.getInt("t"), nbttagcompound5.getInt("p"));
+                    world.scheduleUponLoading(chunk, new BlockPosition(nbttagcompound5.getInt("x"), nbttagcompound5.getInt("y"), nbttagcompound5.getInt("z")), block, nbttagcompound5.getInt("t"), nbttagcompound5.getInt("p")); // Blossom - optimise scheduled block ticking
                 }
             }
         }
diff --git a/src/main/java/net/minecraft/server/NextTickListEntry.java b/src/main/java/net/minecraft/server/NextTickListEntry.java
index 648d25534..2027c5ccb 100644
--- a/src/main/java/net/minecraft/server/NextTickListEntry.java
+++ b/src/main/java/net/minecraft/server/NextTickListEntry.java
@@ -3,7 +3,7 @@ package net.minecraft.server;
 public class NextTickListEntry implements Comparable<NextTickListEntry> {
 
     private static long d;
-    private final Block e;
+    public final Block e; // Blossom - optimise scheduled block ticking
     public final BlockPosition a;
     public long b;
     public int c;
@@ -11,7 +11,13 @@ public class NextTickListEntry implements Comparable<NextTickListEntry> {
 
     public NextTickListEntry(BlockPosition blockposition, Block block) {
         this.f = (long) (NextTickListEntry.d++);
-        this.a = blockposition;
+        // Blossom start - optimise scheduled block ticking
+        if (blockposition instanceof BlockPosition.MutableBlockPosition) {
+            this.a = new BlockPosition(blockposition.getX(), blockposition.getY(), blockposition.getZ());
+        } else {
+            this.a = blockposition;
+        }
+        // Blossom end - optimise scheduled block ticking
         this.e = block;
     }
 
@@ -20,13 +26,17 @@ public class NextTickListEntry implements Comparable<NextTickListEntry> {
             return false;
         } else {
             NextTickListEntry nextticklistentry = (NextTickListEntry) object;
-
-            return this.a.equals(nextticklistentry.a) && Block.a(this.e, nextticklistentry.e);
+            return this.b == nextticklistentry.b && this.c == nextticklistentry.c && this.f == nextticklistentry.f; // Blossom - optimise scheduled block ticking
         }
     }
 
     public int hashCode() {
-        return this.a.hashCode();
+        // Blossom start - optimise scheduled block ticking
+        int result = Long.hashCode(this.b);
+        result = 31 * result + Integer.hashCode(this.c);
+        result = 31 * result + Long.hashCode(this.f);
+        return result;
+        // Blossom end - optimise scheduled block ticking
     }
 
     public NextTickListEntry a(long i) {
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index c05a6a648..0ac4d77ae 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1595,6 +1595,8 @@ public abstract class World implements IBlockAccess {
 
     public void b(BlockPosition blockposition, Block block, int i, int j) {}
 
+    public void scheduleUponLoading(Chunk chunk, BlockPosition blockposition, Block block, int i, int j) {} // Blossom
+
     public void tickEntities() {
         this.methodProfiler.a("entities");
         this.methodProfiler.a("global");
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 9acfa5492..e4ee40d32 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -36,7 +36,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     public EntityTracker tracker;
     private final PlayerChunkMap manager;
     // private final Set<NextTickListEntry> L = Sets.newHashSet(); // PAIL: Rename nextTickListHash
-    private final HashTreeSet<NextTickListEntry> M = new HashTreeSet<NextTickListEntry>(); // CraftBukkit - HashTreeSet // PAIL: Rename nextTickList
+    private final me.samsuik.blossom.world.ScheduledBlockTicks M = new me.samsuik.blossom.world.ScheduledBlockTicks(); // Blossom - optimise scheduled block ticking
     private final Map<UUID, Entity> entitiesByUUID = Maps.newHashMap();
     public ChunkProviderServer chunkProviderServer;
     public boolean savingDisabled;
@@ -48,7 +48,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     private WorldServer.BlockActionDataList[] S = new WorldServer.BlockActionDataList[] { new WorldServer.BlockActionDataList(null), new WorldServer.BlockActionDataList(null)};
     private int T;
     private static final List<StructurePieceTreasure> U = Lists.newArrayList(new StructurePieceTreasure[] { new StructurePieceTreasure(Items.STICK, 0, 1, 3, 10), new StructurePieceTreasure(Item.getItemOf(Blocks.PLANKS), 0, 1, 3, 10), new StructurePieceTreasure(Item.getItemOf(Blocks.LOG), 0, 1, 3, 10), new StructurePieceTreasure(Items.STONE_AXE, 0, 1, 1, 3), new StructurePieceTreasure(Items.WOODEN_AXE, 0, 1, 1, 5), new StructurePieceTreasure(Items.STONE_PICKAXE, 0, 1, 1, 3), new StructurePieceTreasure(Items.WOODEN_PICKAXE, 0, 1, 1, 5), new StructurePieceTreasure(Items.APPLE, 0, 2, 3, 5), new StructurePieceTreasure(Items.BREAD, 0, 2, 3, 3), new StructurePieceTreasure(Item.getItemOf(Blocks.LOG2), 0, 1, 3, 10)});
-    private List<NextTickListEntry> V = Lists.newArrayList();
+    private final ArrayDeque<NextTickListEntry> V = new ArrayDeque<>(); // Blossom - optimise scheduled block ticking
 
     // CraftBukkit start
     public final int dimension;
@@ -559,21 +559,29 @@ public class WorldServer extends World implements IAsyncTaskHandler {
             i = 1;
         }
 
-        if (this.areChunksLoadedBetween(blockposition.a(-b0, -b0, -b0), blockposition.a(b0, b0, b0))) {
+        // Blossom start - optimise scheduled block ticking
+        Chunk chunk = this.getChunkIfLoaded(blockposition);
+        if (chunk != null/*this.areChunksLoadedBetween(blockposition.a(-b0, -b0, -b0), blockposition.a(b0, b0, b0))*/) {
             if (block.getMaterial() != Material.AIR) {
                 nextticklistentry.a((long) i + this.worldData.getTime());
                 nextticklistentry.a(j);
             }
 
             // CraftBukkit - use M, PAIL: Rename nextTickList
-            if (!this.M.contains(nextticklistentry)) {
+            if (chunk.scheduledToTick.add(nextticklistentry)/*!this.M.contains(nextticklistentry)*/) {
                 this.M.add(nextticklistentry);
+                // Blossom end - optimise scheduled block ticking
             }
         }
 
     }
 
     public void b(BlockPosition blockposition, Block block, int i, int j) {
+        // Blossom start - optimise scheduled block ticking
+        scheduleUponLoading(null, blockposition, block, i, j);
+    }
+    public void scheduleUponLoading(Chunk chunk, BlockPosition blockposition, Block block, int i, int j) {
+        // Blossom end - optimise scheduled block ticking
         NextTickListEntry nextticklistentry = new NextTickListEntry(blockposition, block);
 
         nextticklistentry.a(j);
@@ -582,8 +590,10 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         }
 
         // CraftBukkit - use M, PAIL: Rename nextTickList
-        if (!this.M.contains(nextticklistentry)) {
+        // Blossom start - optimise scheduled block ticking
+        if (chunk != null && chunk.scheduledToTick.add(nextticklistentry)/*!this.M.contains(nextticklistentry)*/) {
             this.M.add(nextticklistentry);
+            // Blossom end - optimise scheduled block ticking
         }
 
     }
@@ -635,47 +645,53 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                 timings.scheduledBlocksCleanup.startTiming(); // Spigot
                 NextTickListEntry nextticklistentry;
 
-                for (int j = 0; j < i; ++j) {
-                    nextticklistentry = (NextTickListEntry) this.M.first();
+                // Blossom start - optimise scheduled block ticking
+                Iterator<NextTickListEntry> scheduledBlockItr = this.M.iterator();
+                int count = 0;
+                while (scheduledBlockItr.hasNext()) {
+                    nextticklistentry = scheduledBlockItr.next();
                     if (!flag && nextticklistentry.b > this.worldData.getTime()) {
                         break;
                     }
 
-                    // CraftBukkit - use M, PAIL: Rename nextTickList
-                    this.M.remove(nextticklistentry);
-                    this.V.add(nextticklistentry);
-                }
-                timings.scheduledBlocksCleanup.stopTiming(); // Spigot
+                    boolean tickBlock = count < i;
 
-                // PaperSpigot start - Allow redstone ticks to bypass the tickNextTickListCap
-                if (paperSpigotConfig.tickNextTickListCapIgnoresRedstone) {
-                    Iterator<NextTickListEntry> iterator = this.M.iterator();
-                    while (iterator.hasNext()) {
-                        NextTickListEntry next = iterator.next();
-                        if (!flag && next.b > this.worldData.getTime()) {
-                            break;
-                        }
+                    // PaperSpigot start - Allow redstone ticks to bypass the tickNextTickListCap
+                    if (paperSpigotConfig.tickNextTickListCapIgnoresRedstone) {
+                        Block block = nextticklistentry.a();
+                        tickBlock |= block.isPowerSource() || block instanceof IContainer || block instanceof BlockFalling; // Blossom - falling blocks should bypass tick cap
+                    } else if (!tickBlock) {
+                        break;
+                    }
+                    // PaperSpigot end - Allow redstone ticks to bypass the tickNextTickListCap
 
-                        if (next.a().isPowerSource() || next.a() instanceof IContainer || next.a() instanceof BlockFalling) { // Blossom - falling blocks should bypass tick cap
-                            iterator.remove();
-                            this.V.add(next);
+                    if (tickBlock) {
+                        Chunk chunk = this.getChunkIfLoaded(nextticklistentry.a);
+                        if (chunk != null) {
+                            chunk.scheduledToTick.remove(nextticklistentry);
                         }
+                        scheduledBlockItr.remove();
+                        this.V.add(nextticklistentry);
+                        count++;
                     }
                 }
-                // PaperSpigot end
+                // Blossom end - optimise scheduled block ticking
+                timings.scheduledBlocksCleanup.stopTiming(); // Spigot
+
+                // Blossom - optimise scheduled block ticking; moved up
 
                 this.methodProfiler.b();
                 this.methodProfiler.a("ticking");
                 timings.scheduledBlocksTicking.startTiming(); // Spigot
-                Iterator iterator = this.V.iterator();
 
-                while (iterator.hasNext()) {
-                    nextticklistentry = (NextTickListEntry) iterator.next();
-                    iterator.remove();
-                    byte b0 = 0;
-
-                    if (this.areChunksLoadedBetween(nextticklistentry.a.a(-b0, -b0, -b0), nextticklistentry.a.a(b0, b0, b0))) {
-                        IBlockData iblockdata = this.getType(nextticklistentry.a);
+                // Blossom start - optimise scheduled block ticking
+                while ((nextticklistentry = this.V.poll()) != null) {
+                    BlockPosition position = nextticklistentry.a;
+                    Chunk chunk = this.getChunkIfLoaded(position.getX() >> 4, position.getZ() >> 4);
+                    if (chunk != null/*this.areChunksLoadedBetween(nextticklistentry.a.a(-b0, -b0, -b0), nextticklistentry.a.a(b0, b0, b0))*/) {
+                        //chunk.scheduledToTick.remove(nextticklistentry);
+                        IBlockData iblockdata = chunk.getBlockData(nextticklistentry.a);
+                        // Blossom end - optimise scheduled block ticking
                         co.aikar.timings.Timing timing = iblockdata.getBlock().getTiming(); // Spigot
                         timing.startTiming(); // Spigot
 
@@ -714,7 +730,58 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         return this.a(new StructureBoundingBox(i, 0, k, j, 256, l), flag);
     }
 
+    // Blossom start - optimise nextticklist unloading
+    private List<NextTickListEntry> getScheduledBlocksInsideBoundingBox(
+        final StructureBoundingBox boundingBox,
+        final int minChunkX,
+        final int minChunkZ,
+        final int maxChunkX,
+        final int maxChunkZ,
+        final boolean remove
+    ) {
+        List<NextTickListEntry> scheduledTickingBlocks = new it.unimi.dsi.fastutil.objects.ReferenceArrayList<>();
+        for (int chunkX = minChunkX; chunkX <= maxChunkX; ++chunkX) {
+            for (int chunkZ = minChunkZ; chunkZ <= maxChunkZ; ++chunkZ) {
+                Chunk chunk = this.getChunkIfLoaded(chunkX, chunkZ);
+                if (chunk == null) {
+                    continue;
+                }
+
+                for (Iterator<NextTickListEntry> itr = chunk.scheduledToTick.iterator(); itr.hasNext(); ) {
+                    NextTickListEntry scheduledBlock = itr.next();
+                    BlockPosition position = scheduledBlock.a;
+
+                    if (position.getX() >= boundingBox.a && position.getX() < boundingBox.d && position.getZ() >= boundingBox.c && position.getZ() < boundingBox.f) {
+                        if (remove) {
+                            itr.remove();
+                        }
+                        scheduledTickingBlocks.add(scheduledBlock);
+                    }
+                }
+            }
+        }
+
+        if (remove) {
+            // Might be worth adding a flag to NextTickListEntry to indicate it shouldn't be ticked.
+            for (NextTickListEntry scheduledBlock : scheduledTickingBlocks) {
+                this.M.remove(scheduledBlock);
+                this.V.remove(scheduledBlock);
+            }
+        }
+
+        return scheduledTickingBlocks;
+    }
+
     public List<NextTickListEntry> a(StructureBoundingBox structureboundingbox, boolean flag) {
+        int minChunkX = structureboundingbox.a >> 4;
+        int minChunkZ = structureboundingbox.c >> 4;
+        int maxChunkX = structureboundingbox.d >> 4;
+        int maxChunkZ = structureboundingbox.f >> 4;
+        int totalChunks = (maxChunkX - minChunkX + 1) * (maxChunkZ - minChunkZ + 1);
+        if (totalChunks < 16 * 16) {
+            return this.getScheduledBlocksInsideBoundingBox(structureboundingbox, minChunkX, minChunkZ, maxChunkX, maxChunkZ, flag);
+        }
+        // Blossom end - optimise scheduled block ticking
         ArrayList arraylist = null;
 
         for (int i = 0; i < 2; ++i) {
-- 
2.50.1 (Apple Git-155)

