From b773cc53e6ed94338137af56166686aa241e25ca Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 27 May 2023 19:52:00 +0100
Subject: [PATCH] Scoreboard Objective Updates


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index cdd0778ff..2c9a8609c 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -343,4 +343,9 @@ public final class BlossomConfig {
         batchExplosions = getBoolean("explosions.batch-explosions", true);
     }
 
+    public static boolean playerObjectiveUpdates;
+    private static void PlayerObjectiveUpdates() {
+        playerObjectiveUpdates = getBoolean("player.scoreboard-objective-updates", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index c28b5a57b..1c347f70c 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -847,7 +847,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
 
     public void a(Statistic statistic, int i) {
-        if (statistic != null) {
+        if (statistic != null && me.samsuik.blossom.configuration.BlossomConfig.playerObjectiveUpdates) { // Blossom
             this.bK.b(this, statistic, i);
             Iterator iterator = this.getScoreboard().getObjectivesForCriteria(statistic.k()).iterator();
 
-- 
2.50.1 (Apple Git-155)

