From 5af652026ff2606e010af01ec07d7d6796712e9a Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 2 Jul 2022 15:15:50 +0100
Subject: [PATCH] Generic Entity Optimisations


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index e2d649132..6b433103b 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -861,7 +861,7 @@ public abstract class Entity implements ICommandListener {
                 this.N = (float) ((double) this.N + (double) MathHelper.sqrt(d21 * d21 + d22 * d22 + d23 * d23) * 0.6D);
                 if (this.N > (float) this.h && block.getMaterial() != Material.AIR) {
                     this.h = (int) this.N + 1;
-                    if (this.V()) {
+                    if (false && this.V()) { // Blossom
                         float f = MathHelper.sqrt(this.motX * this.motX * 0.20000000298023224D + this.motY * this.motY + this.motZ * this.motZ * 0.20000000298023224D) * 0.35F;
 
                         if (f > 1.0F) {
@@ -967,7 +967,7 @@ public abstract class Entity implements ICommandListener {
     protected void a(BlockPosition blockposition, Block block) {
         Block.StepSound block_stepsound = block.stepSound;
 
-        if (this.world.getType(blockposition.up()).getBlock() == Blocks.SNOW_LAYER) {
+        if (getChunkAtLocation().getBlockData(blockposition.up()).getBlock() == Blocks.SNOW_LAYER) { // Blossom
             block_stepsound = Blocks.SNOW_LAYER.stepSound;
             this.makeSound(block_stepsound.getStepSound(), block_stepsound.getVolume1() * 0.15F, block_stepsound.getVolume2());
         } else if (!block.getMaterial().isLiquid()) {
@@ -1059,6 +1059,7 @@ public abstract class Entity implements ICommandListener {
     }
 
     protected void X() {
+        if (true) return; // Blossom
         float f = MathHelper.sqrt(this.motX * this.motX * 0.20000000298023224D + this.motY * this.motY + this.motZ * this.motZ * 0.20000000298023224D) * 0.2F;
 
         if (f > 1.0F) {
@@ -1087,7 +1088,7 @@ public abstract class Entity implements ICommandListener {
     }
 
     public void Y() {
-        if (this.isSprinting() && !this.V()) {
+        if (false && this.isSprinting() && !this.V()) { // Blossom
             this.Z();
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 8cc0cef0b..ad2a14e63 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -224,7 +224,7 @@ public abstract class EntityLiving extends Entity {
             }
         }
 
-        if (this.isAlive() && this.U()) {
+        if (this.fireTicks != 0 && this.isAlive() && this.U()) { // Blossom
             this.extinguish();
         }
 
@@ -470,6 +470,7 @@ public abstract class EntityLiving extends Entity {
     // CraftBukkit end
 
     protected void bi() {
+        if (this.effects.isEmpty() && this.effectsToProcess.isEmpty()) return; // Blossom
         Iterator iterator = this.effects.keySet().iterator();
 
         isTickingEffects = true; // CraftBukkit
-- 
2.40.0.windows.1

