From 4ba5ef333481d5d4bfacb2b1c15eef9b4040ffec Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 11 Jun 2023 00:17:24 +0100
Subject: [PATCH] Fix block break desync


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index eb569576a..c8c7636e9 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -642,6 +642,12 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
             if (d3 > 36.0D) {
+                // Blossom start - fix block break desync
+                int maxDistance = Math.max(player.viewDistance / 2, 1);
+                if (d3 < Math.pow(maxDistance, 2)) {
+                    this.player.playerConnection.sendPacket(new PacketPlayOutBlockChange(worldserver, blockposition));
+                }
+                // Blossom end
                 return;
             } else if (blockposition.getY() >= this.minecraftServer.getMaxBuildHeight()) {
                 return;
-- 
2.50.1 (Apple Git-155)

