From 08550132106fdb5dafdeacd2b7c3d7e2a4cee00a Mon Sep 17 00:00:00 2001
From: "40902469+Samsuik@users.noreply.github.com"
 <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 19 Dec 2020 19:35:00 +0000
Subject: [PATCH] Configurable Golems


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 57e46cf51..25d20ae25 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -324,4 +324,11 @@ public class BlossomConfig {
         potionVerticalOffset = (float) getDouble("projectiles.potion.vertical-offset", -20.0);
     }
 
+    public static boolean ironGolemsDropPoppies;
+    public static boolean ironGolemsFallDamage;
+    private static void IronGolems() {
+        ironGolemsDropPoppies = getBoolean("mobs.iron-golems.drop-poppies", true);
+        ironGolemsFallDamage = getBoolean("mobs.iron-golems.take-fall-damage", false);
+    }
+
 }
-- 
2.40.0.windows.1

