From 307c504730295362cf689b500bbad16e5d7cd7da Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 11 Jun 2023 15:04:21 +0100
Subject: [PATCH] Limit players that can see block breaking


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 920872a3f..4c6d46590 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -3534,12 +3534,20 @@ public abstract class World implements IBlockAccess {
     }
 
     public void c(int i, BlockPosition blockposition, int j) {
-        for (int k = 0; k < this.u.size(); ++k) {
-            IWorldAccess iworldaccess = (IWorldAccess) this.u.get(k);
+        AxisAlignedBB bb = new AxisAlignedBB(blockposition, blockposition).grow(32, 32, 32);
+        List<EntityPlayer> players = getEntitiesLimitedByType(EntityPlayer.class, bb, null, 6, 16);
 
-            iworldaccess.b(i, blockposition, j);
-        }
+        for (EntityPlayer entityplayer : players) {
+            if (entityplayer.getId() != i) {
+                double d0 = (double) blockposition.getX() - entityplayer.locX;
+                double d1 = (double) blockposition.getY() - entityplayer.locY;
+                double d2 = (double) blockposition.getZ() - entityplayer.locZ;
 
+                if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0) {
+                    entityplayer.playerConnection.sendPacket(new PacketPlayOutBlockBreakAnimation(i, blockposition, j));
+                }
+            }
+        }
     }
 
     public Calendar Y() {
-- 
2.50.1

