From 25c2a966edaf2079bf687d97b3ce2411f294f4bc Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 27 May 2023 18:42:58 +0100
Subject: [PATCH] Ensure entity is in range before checking hasLineOfSight


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 315492671..793c80219 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1592,14 +1592,10 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
 
         this.player.resetIdleTimer();
         if (entity != null) {
-            boolean flag = this.player.hasLineOfSight(entity);
-            double d0 = 36.0D;
-
-            if (!flag) {
-                d0 = 9.0D;
-            }
-
-            if (this.player.h(entity) < d0) {
+            // Blossom start - we need to check the distance first otherwise hasLineOfSight can be abused
+            double distanceSquared = this.player.h(entity);
+            if (distanceSquared < 36.0D && (this.player.hasLineOfSight(entity) || distanceSquared < 9.0D)) {
+                // Blossom end
                 ItemStack itemInHand = this.player.inventory.getItemInHand(); // CraftBukkit
 
                 if (packetplayinuseentity.a() == PacketPlayInUseEntity.EnumEntityUseAction.INTERACT
-- 
2.40.0.windows.1

