From a8698f2ff33c893f946df3350cd1e8ef0d659374 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 11 Jun 2023 16:55:53 +0100
Subject: [PATCH] Lag compensated block breaking


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 102e6bc7d..fb07f8a17 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -80,6 +80,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
     // Spigot end
     public final boolean realPlayer; // Blossom - optimise entity tracker
+    public int movementTick = 0; // Blossom - lag compensated block breaking
 
     public EntityPlayer(MinecraftServer minecraftserver, WorldServer worldserver, GameProfile gameprofile, PlayerInteractManager playerinteractmanager) {
         super(worldserver, gameprofile);
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 11e5a7fc6..6d6303654 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -432,6 +432,7 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
 
             this.h = true;
             if (!this.player.viewingCredits) {
+                this.player.movementTick++; // Blossom - lag compensated block breaking
                 double d0 = this.player.locX;
                 double d1 = this.player.locY;
                 double d2 = this.player.locZ;
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index 404641a2f..f941b7da6 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -15,6 +15,8 @@ public class PlayerInteractManager {
     private WorldSettings.EnumGamemode gamemode;
     private boolean d;
     private int lastDigTick;
+    private int lastPlayerDigTick; // Blossom
+    private long startBreaking; // Blossom
     private BlockPosition f;
     private int currentTick;
     private boolean h;
@@ -144,6 +146,8 @@ public class PlayerInteractManager {
 
             // this.world.douseFire((EntityHuman) null, blockposition, enumdirection); // CraftBukkit - Moved down
             this.lastDigTick = this.currentTick;
+            this.lastPlayerDigTick = this.player.movementTick; // Blossom - lag compensated block breaking
+            this.startBreaking = System.currentTimeMillis(); // Blossom - lag compensated block breaking
             float f = 1.0F;
 
             // CraftBukkit start - Swings at air do *NOT* exist.
@@ -208,9 +212,16 @@ public class PlayerInteractManager {
             Block block = this.world.getType(blockposition).getBlock();
 
             if (block.getMaterial() != Material.AIR) {
-                float f = block.getDamage(this.player, this.player.world, blockposition) * (float) (i + 1);
+                // Blossom start - lag compensated block breaking
+                long timeInTicks = (System.currentTimeMillis() - this.startBreaking) / 50L;
+                int playerDigTicks = this.player.movementTick - this.lastPlayerDigTick;
+                float damagePerTick = block.getDamage(this.player, this.player.world, blockposition);
 
-                if (f >= 0.7F) {
+                float requiredTicks = 1f / damagePerTick;
+                float compensation = 20f / Math.min((float) MinecraftServer.getServer().tps1.getAverage(), 20f) - 1f;
+
+                if (timeInTicks + 1 >= requiredTicks - compensation || playerDigTicks >= requiredTicks) {
+                    // Blossom end - lag compensated block breaking
                     this.d = false;
                     this.world.c(this.player.getId(), blockposition, -1);
                     this.breakBlock(blockposition);
-- 
2.50.1

