From 4029da4340fb753aa28733c97adc57897ad0a8bd Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 16 Jun 2023 20:33:09 +0100
Subject: [PATCH] Activation Range Delay


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index f5dacad4d..be2373fe7 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -375,4 +375,9 @@ public class BlossomConfig {
         mergeEntitiesOnSpawn = getBoolean("tnt.merging.on-spawn", true);
     }
 
+    public static int activationRangeDelay;
+    private static void ActivationRange() {
+        activationRangeDelay = Math.max(getInt("mobs.activation-range-delay", 0) + 1, 1);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 745b53bed..5dbda2e8d 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1646,6 +1646,7 @@ public abstract class World implements IBlockAccess {
         timings.entityRemoval.stopTiming(); // Spigot
         this.methodProfiler.c("regular");
 
+        if (MinecraftServer.currentTick % me.samsuik.blossom.configuration.BlossomConfig.activationRangeDelay == 0) // Blossom
         org.spigotmc.ActivationRange.activateEntities(this); // Spigot
         timings.entityTick.startTiming(); // Spigot
         guardEntityList = true; // Spigot
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index 4de43e68e..61071ecb8 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -268,7 +268,7 @@ public class ActivationRange
                 if ( checkEntityImmunities( entity ) )
                 {
                     // Triggered some sort of immunity, give 20 full ticks before we check again.
-                    entity.activatedTick = MinecraftServer.currentTick + 20;
+                    entity.activatedTick = MinecraftServer.currentTick + 20 + me.samsuik.blossom.configuration.BlossomConfig.activationRangeDelay; // Blossom
                 }
                 isActive = true;
             }
-- 
2.50.1 (Apple Git-155)

