From 63beecd30a7a12c3b17dde33b8b96e62d5580ab0 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 19 Jun 2023 18:57:22 +0100
Subject: [PATCH] Defence Patches


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index a65306dc9..05f19e97c 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -387,4 +387,17 @@ public class BlossomConfig {
         optimisedMovement = getBoolean("tnt.optimised-movement", true);
     }
 
+    public static boolean fixSlimeBlockDefence;
+    public static boolean fixCobwebGtDefence;
+    public static boolean sandReplaceNonSolids;
+    public static boolean disableLeftShooting;
+    public static double leftShootingThreshold;
+    private static void Defences() {
+        fixSlimeBlockDefence = getBoolean("tnt.fix-slime-block-defence", false);
+        fixCobwebGtDefence = getBoolean("tnt.fix-cobweb-movement-defence", false);
+        sandReplaceNonSolids = getBoolean("sand.replace-non-solid-blocks", false);
+        disableLeftShooting = getBoolean("tnt.disable-left-shooting", false);
+        leftShootingThreshold = getDouble("tnt.left-shooting-threshold", 80.0);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/BlockSlime.java b/src/main/java/net/minecraft/server/BlockSlime.java
index 331bafa03..83da9eb38 100644
--- a/src/main/java/net/minecraft/server/BlockSlime.java
+++ b/src/main/java/net/minecraft/server/BlockSlime.java
@@ -18,7 +18,7 @@ public class BlockSlime extends BlockHalfTransparent {
     }
 
     public void a(World world, Entity entity) {
-        if (entity.isSneaking()) {
+        if (entity.isSneaking() || me.samsuik.blossom.configuration.BlossomConfig.fixSlimeBlockDefence && entity.isCannoningEntity) { // Blossom
             super.a(world, entity);
         } else if (entity.motY < 0.0D) {
             entity.motY = -entity.motY;
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index c157fe976..18064ce55 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -686,7 +686,7 @@ public abstract class Entity implements ICommandListener {
         }
         // CraftBukkit end
 
-        if (this.H) {
+        if (this.H && (!me.samsuik.blossom.configuration.BlossomConfig.fixCobwebGtDefence || Math.pow(moveX, 2.0) + Math.pow(moveY, 2.0) + Math.pow(moveZ, 2.0) < 48.0 * 48.0)) { // Blossom
             this.H = false;
             moveX *= 0.25D;
             moveY *= 0.05000000074505806D;
diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
index 34b6f6bd3..66f541c27 100644
--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
@@ -84,6 +84,21 @@ public class EntityFallingBlock extends Entity {
         }
     }
     // Blossom end
+    // Blossom start
+    public void preventLeftShooting() {
+        double changeX = Math.abs(motX) - Math.abs(lastMotX);
+        double changeZ = Math.abs(motZ) - Math.abs(lastMotZ);
+        double travelX = Math.abs(sourceLoc.getX() - locX);
+        double travelZ = Math.abs(sourceLoc.getZ() - locZ);
+
+        if (changeX > 0.0 && travelZ > me.samsuik.blossom.configuration.BlossomConfig.leftShootingThreshold) {
+            motX = 0.0;
+        }
+        if (changeZ > 0.0 && travelX > me.samsuik.blossom.configuration.BlossomConfig.leftShootingThreshold) {
+            motZ = 0.0;
+        }
+    }
+    // Blossom end
 
     public void t_() {
         Block block = this.block.getBlock();
@@ -96,6 +111,12 @@ public class EntityFallingBlock extends Entity {
             this.lastZ = this.locZ;
             BlockPosition blockposition;
 
+            // Blossom start
+            if (me.samsuik.blossom.configuration.BlossomConfig.disableLeftShooting && (motX != 0.0 || motZ != 0.0)) {
+                preventLeftShooting();
+            }
+            // Blossom end
+
             if (this.ticksLived++ == 0) {
                 blockposition = new BlockPosition(this);
                 if (this.world.getType(blockposition).getBlock() == block && !CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) {
@@ -138,7 +159,7 @@ public class EntityFallingBlock extends Entity {
                     if (this.world.getType(blockposition).getBlock() != Blocks.PISTON_EXTENSION) {
                         this.die();
                         if (!this.e) {
-                            if (this.world.a(block, blockposition, true, EnumDirection.UP, (Entity) null, (ItemStack) null) && !BlockFalling.canFall(this.world, blockposition.down()) /* mimic the false conditions of setTypeIdAndData */ && blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000 && blockposition.getY() >= 0 && blockposition.getY() < (this.world.tacoSpigotConfig.disableFallingBlockStackingAt256 ? 255 : 256) && this.world.getType(blockposition) != this.block) {
+                            if (this.world.isObstructed(block, blockposition, true, EnumDirection.UP, (Entity) null, (ItemStack) null, me.samsuik.blossom.configuration.BlossomConfig.sandReplaceNonSolids) && !BlockFalling.canFall(this.world, blockposition.down()) /* mimic the false conditions of setTypeIdAndData */ && blockposition.getX() >= -30000000 && blockposition.getZ() >= -30000000 && blockposition.getX() < 30000000 && blockposition.getZ() < 30000000 && blockposition.getY() >= 0 && blockposition.getY() < (this.world.tacoSpigotConfig.disableFallingBlockStackingAt256 ? 255 : 256) && this.world.getType(blockposition) != this.block) { // Blossom
                                 if (CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this.block.getBlock(), this.block.getBlock().toLegacyData(this.block)).isCancelled()) {
                                     return;
                                 }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 6dea49273..ba3692128 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -3275,11 +3275,16 @@ public abstract class World implements IBlockAccess {
     }
 
     public boolean a(Block block, BlockPosition blockposition, boolean flag, EnumDirection enumdirection, Entity entity, ItemStack itemstack) {
+        // Blossom start
+        return isObstructed(block, blockposition, flag, enumdirection, entity, itemstack, false);
+    }
+    public boolean isObstructed(Block block, BlockPosition blockposition, boolean flag, EnumDirection enumdirection, Entity entity, ItemStack itemstack, boolean replaceNotSolid) {
         Block block1 = this.getType(blockposition).getBlock();
         AxisAlignedBB axisalignedbb = flag ? null : block.a(this, blockposition, block.getBlockData());
 
         // CraftBukkit start - store default return
-        boolean defaultReturn = axisalignedbb != null && !this.a(axisalignedbb, entity) ? false : (block1.getMaterial() == Material.ORIENTABLE && block == Blocks.ANVIL ? true : block1.getMaterial().isReplaceable() && block.canPlace(this, blockposition, enumdirection, itemstack));
+        boolean defaultReturn = axisalignedbb != null && !this.a(axisalignedbb, entity) ? false : (block1.getMaterial() == Material.ORIENTABLE && block == Blocks.ANVIL ? true : (replaceNotSolid && !block1.getMaterial().isSolid() || block1.getMaterial().isReplaceable()) && block.canPlace(this, blockposition, enumdirection, itemstack));
+        // Blossom end
         BlockCanBuildEvent event = new BlockCanBuildEvent(this.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), CraftMagicNumbers.getId(block), defaultReturn);
         this.getServer().getPluginManager().callEvent(event);
 
-- 
2.50.1

