From 52942bc15eeb046f09c692a595a51db79d7d3aba Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Wed, 21 Jun 2023 18:20:50 +0100
Subject: [PATCH] Disallow projectiles moving through portals


diff --git a/src/main/java/net/minecraft/server/EntityProjectile.java b/src/main/java/net/minecraft/server/EntityProjectile.java
index 2f757ee38..ad1bd8f59 100644
--- a/src/main/java/net/minecraft/server/EntityProjectile.java
+++ b/src/main/java/net/minecraft/server/EntityProjectile.java
@@ -163,7 +163,9 @@ public abstract class EntityProjectile extends Entity implements IProjectile {
 
         if (movingobjectposition != null) {
             if (movingobjectposition.type == MovingObjectPosition.EnumMovingObjectType.BLOCK && this.world.getType(movingobjectposition.a()).getBlock() == Blocks.PORTAL) {
-                this.d(movingobjectposition.a());
+                // Blossom start - don't allow projectiles through portals
+                // this.d(movingobjectposition.a());
+                // Blossom end
             } else {
                 // Blossom start
                 boolean hit = true;
-- 
2.40.0.windows.1

