From 1acecd169df3d21d9fab0141a1ba9b2d0b824786 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 2 Jul 2023 00:34:24 +0100
Subject: [PATCH] Heavily Optimise Game Registries


diff --git a/src/main/java/net/minecraft/server/Block.java b/src/main/java/net/minecraft/server/Block.java
index f9d63d580..a17e07910 100644
--- a/src/main/java/net/minecraft/server/Block.java
+++ b/src/main/java/net/minecraft/server/Block.java
@@ -110,6 +110,23 @@ public class Block {
     protected final BlockStateList blockStateList;
     private IBlockData blockData;
     private String name;
+    // Blossom start
+    private int id = -1;
+
+    public final int getId() {
+        return id;
+    }
+
+    @Override
+    public final int hashCode() {
+        return id;
+    }
+
+    @Override
+    public final boolean equals(Object object) {
+        return this == object;
+    }
+    // Blossom end
 
     public static int getId(Block block) {
         return Block.REGISTRY.b(block);
@@ -1113,6 +1130,7 @@ public class Block {
     }
 
     private static void a(int i, MinecraftKey minecraftkey, Block block) {
+        block.id = i; // Blossom
         Block.REGISTRY.a(i, minecraftkey, block);
     }
 
diff --git a/src/main/java/net/minecraft/server/BlockStateList.java b/src/main/java/net/minecraft/server/BlockStateList.java
index 61dba6cab..76ad0732d 100644
--- a/src/main/java/net/minecraft/server/BlockStateList.java
+++ b/src/main/java/net/minecraft/server/BlockStateList.java
@@ -116,6 +116,7 @@ public class BlockStateList {
 
     static class BlockData extends BlockDataAbstract {
 
+        private static int UNIQUE_COUNTER = 0; // Blossom
         private final Block a;
         // TacoSpigot start
         private final ImmutableMap<IBlockState, Comparable> bAsImmutableMap;
@@ -128,6 +129,7 @@ public class BlockStateList {
         private boolean dynamic = true;
         private boolean collideAsFullBlock = false;
         // Blossom end
+        private final int id = UNIQUE_COUNTER++; // Blossom
 
         private BlockData(Block block, ImmutableMap<IBlockState, Comparable> immutablemap) {
             this.a = block;
@@ -232,7 +234,7 @@ public class BlockStateList {
         }
 
         public int hashCode() {
-            return this.b.hashCode();
+            return id; // Blossom
         }
 
         public void a(Map<Map<IBlockState, Comparable>, BlockStateList.BlockData> map) {
diff --git a/src/main/java/net/minecraft/server/Item.java b/src/main/java/net/minecraft/server/Item.java
index 87a2d7b90..8ee5e4b10 100644
--- a/src/main/java/net/minecraft/server/Item.java
+++ b/src/main/java/net/minecraft/server/Item.java
@@ -11,7 +11,7 @@ import java.util.UUID;
 public class Item {
 
     public static final RegistryMaterials<MinecraftKey, Item> REGISTRY = new RegistryMaterials();
-    private static final Map<Block, Item> a = Maps.newHashMap();
+    private static final Map<Block, Item> a = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>(); // Blossom
     protected static final UUID f = UUID.fromString("CB3F55D3-645C-4F38-A497-9C13A33DB5CF");
     private CreativeModeTab b;
     protected static Random g = new Random();
@@ -22,6 +22,23 @@ public class Item {
     private Item craftingResult;
     private String k;
     private String name;
+    // Blossom start
+    private int id = -1;
+
+    public final int getId() {
+        return id;
+    }
+
+    @Override
+    public final int hashCode() {
+        return id;
+    }
+
+    @Override
+    public final boolean equals(Object object) {
+        return this == object;
+    }
+    // Blossom end
 
     public Item() {}
 
@@ -738,6 +755,7 @@ public class Item {
     }
 
     private static void a(int i, MinecraftKey minecraftkey, Item item) {
+        item.id = i; // Blossom
         Item.REGISTRY.a(i, minecraftkey, item);
     }
 
diff --git a/src/main/java/net/minecraft/server/RegistryID.java b/src/main/java/net/minecraft/server/RegistryID.java
index d9d458239..c3621cede 100644
--- a/src/main/java/net/minecraft/server/RegistryID.java
+++ b/src/main/java/net/minecraft/server/RegistryID.java
@@ -9,32 +9,73 @@ import java.util.List;
 
 public class RegistryID<T> implements Registry<T> {
 
-    private final IdentityHashMap<T, Integer> a = new IdentityHashMap(512);
-    private final List<T> b = Lists.newArrayList();
+    // Blossom start - heavily optimise game registries
+    private char[] indexes = new char[0];
+    private T[] storage = (T[]) new Object[0];
+
+    public final T getRaw(final char id) {
+        return this.storage[id];
+    }
+    // Blossom end - heavily optimise game registries
 
     public RegistryID() {}
 
     public void a(T t0, int i) {
-        this.a.put(t0, Integer.valueOf(i));
+        // Blossom start - heavily optimise game registries
+        final int oldId = i;
+        if (i < 0 && (i = Character.MAX_VALUE + i) >= 0) {
+            MinecraftServer.LOGGER.warn("negative index added to registry was remapped ({} -> {})", oldId, i);
+        }
 
-        while (this.b.size() <= i) {
-            this.b.add(null);
+        assert t0 != null : String.format("null entries cannot be added to this registry (%d)", i);
+        assert i >= 0 : String.format("negative index was unable to be remapped (%d)", i);
+        assert i < Character.MAX_VALUE : String.format("(%d<>%d, %s) index is greater than 327677", t0.hashCode(), i, t0);
+
+        // ensure storage and index capacity
+        if (i >= this.storage.length) {
+            this.storage = java.util.Arrays.copyOf(this.storage, i + 1);
         }
 
-        this.b.set(i, t0);
+        final int hash = t0.hashCode();
+        if (hash >= this.indexes.length) {
+            this.indexes = java.util.Arrays.copyOf(this.indexes, hash + 1);
+        }
+
+        // add to registry
+        this.storage[i] = t0;
+        this.indexes[hash] = (char) i;
+        // Blossom end - heavily optimise game registries
     }
 
     public int b(T t0) {
-        Integer integer = (Integer) this.a.get(t0);
-
-        return integer == null ? -1 : integer.intValue();
+        return t0 == null ? -1 : this.indexes[t0.hashCode()]; // Blossom - heavily optimise game registries
     }
 
     public final T a(int i) {
-        return i >= 0 && i < this.b.size() ? this.b.get(i) : null;
+        return i >= 0 && i < this.storage.length ? this.storage[i] : null; // Blossom - heavily optimise game registries
     }
 
     public Iterator<T> iterator() {
-        return Iterators.filter(this.b.iterator(), Predicates.notNull());
+        // Blossom start - heavily optimise game registries
+        return new Iterator<T>() {
+            private int index = -1;
+
+            @Override
+            public boolean hasNext() {
+                final T[] storage = RegistryID.this.storage;
+                while (++this.index < storage.length) {
+                    if (storage[index] != null) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            @Override
+            public T next() {
+                return RegistryID.this.storage[index];
+            }
+        };
+        // Blossom end - heavily optimise game registries
     }
 }
-- 
2.50.1

