From df7a06a5435fadebef3a207281e12d27945977bc Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 10 Jul 2023 00:23:53 +0100
Subject: [PATCH] Fix Paper player-blocking-damage-multiplier


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 2f9aab57b..80de90003 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1094,7 +1094,7 @@ public abstract class EntityLiving extends Entity {
                 public Double apply(Double f) {
                     if (human) {
                         if (!damagesource.ignoresArmor() && ((EntityHuman) EntityLiving.this).isBlocking() && f > 0.0F) {
-                            return -(f - ((1.0F + f) * 0.5F));
+                            return -(f - ((1.0F + f) * EntityLiving.this.world.paperSpigotConfig.playerBlockingDamageMultiplier)); // Blossom - seriously?
                         }
                     }
                     return -0.0;
-- 
2.50.1

