From 2b38404c2a1149cd5784e55eddff4a18f8a977e0 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sun, 9 Jul 2023 20:54:30 +0100
Subject: [PATCH] Fix getChunkAtLocation incorrect chunk position


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 4c6b58e8e..1a21f6baf 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -327,7 +327,7 @@ public abstract class Entity implements ICommandListener {
      * Will only return null if the location specified is not loaded
      */
     public Chunk getChunkAtLocation() {
-        return getCurrentChunkAt((int)Math.floor(locX) >> 4, (int)Math.floor(locZ) >> 4);
+        return getCurrentChunkAt(MathHelper.floor(locX) >> 4, MathHelper.floor(locZ) >> 4); // Blossom - incorrect chunk position
     }
     // Paper end
     // Paper start - make end portalling safe
-- 
2.49.0

