From 84005918b32d6a3093c02fd2e47995ab9c1b1f36 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 18 Jul 2023 23:32:01 +0100
Subject: [PATCH] Actual Throttle Placement Configuration

smh, paper

diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 148993018..39aee934e 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -558,4 +558,9 @@ public final class BlossomConfig {
         chunkLoadingPlayersPerThread = getInt("chunk.loading.players-per-thread", 50);
     }
 
+    public static boolean throttlePlacement;
+    private static void ThrottlePlacement() {
+        throttlePlacement = getBoolean("player.throttle-block-placement", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index b0d69085f..b0b36af55 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -709,6 +709,11 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             packets = 0;
         }
     // Spigot end
+        // Blossom start
+        if (!me.samsuik.blossom.configuration.BlossomConfig.throttlePlacement) {
+            lastPlace = -1;
+        }
+        // Blossom end
 
         // CraftBukkit start
         if (this.player.dead) return;
-- 
2.50.1 (Apple Git-155)

