From a40202542c1c6b98a620c33458a7e5d552f80cd0 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 10 Jul 2023 17:24:45 +0100
Subject: [PATCH] Dispenser Item Selection Randomness


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index e02bf2943..2c94c894d 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -463,5 +463,10 @@ public class BlossomConfig {
         mobHitDelay = getInt("mobs.hit-delay", 10);
         playerHitDifference = getBoolean("player.allow-hit-difference", true);
     }
+
+    public static boolean dispenserRandomness;
+    private static void DispenserRandomness() {
+        dispenserRandomness = getBoolean("dispensers.items-selection-randomness", true);
+    }
     
 }
diff --git a/src/main/java/net/minecraft/server/TileEntityDispenser.java b/src/main/java/net/minecraft/server/TileEntityDispenser.java
index 9be494ba5..83571bd6d 100644
--- a/src/main/java/net/minecraft/server/TileEntityDispenser.java
+++ b/src/main/java/net/minecraft/server/TileEntityDispenser.java
@@ -89,7 +89,7 @@ public class TileEntityDispenser extends TileEntityContainer implements IInvento
         int j = 1;
 
         for (int k = 0; k < this.items.length; ++k) {
-            if (this.items[k] != null && TileEntityDispenser.f.nextInt(j++) == 0) {
+            if (this.items[k] != null && (!me.samsuik.blossom.configuration.BlossomConfig.dispenserRandomness || TileEntityDispenser.f.nextInt(j++) == 0)) { // Blossom
                 if (this.items[k].count == 0) continue; // CraftBukkit
                 i = k;
             }
-- 
2.40.0.windows.1

