From b45ebaad93fb33ac47c7bca4f9921a7e725dbbd2 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 15 Jul 2023 17:29:49 +0100
Subject: [PATCH] Cache Chunk Retrieval


diff --git a/src/main/java/me/samsuik/blossom/chunk/ServerChunkMap.java b/src/main/java/me/samsuik/blossom/chunk/ServerChunkMap.java
new file mode 100644
index 000000000..23282b120
--- /dev/null
+++ b/src/main/java/me/samsuik/blossom/chunk/ServerChunkMap.java
@@ -0,0 +1,25 @@
+package me.samsuik.blossom.chunk;
+
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import net.minecraft.server.Chunk;
+import org.bukkit.craftbukkit.util.LongHash;
+
+public final class ServerChunkMap extends Long2ObjectOpenHashMap<Chunk> {
+    private final Chunk[] recentChunks = new Chunk[8];
+
+    public ServerChunkMap(int expected, float factor) {
+        super(expected, factor);
+    }
+
+    @Override
+    public Chunk get(long chunkKey) {
+        int x = LongHash.msw(chunkKey);
+        int z = LongHash.lsw(chunkKey);
+        int index = ((x << 2) | z & 3) & 7;
+        Chunk chunk = this.recentChunks[index];
+        if (chunk != null && chunk.chunkKey == chunkKey && chunk.isLoaded()) {
+            return chunk;
+        }
+        return this.recentChunks[index] = super.get(chunkKey);
+    }
+}
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index b9a11edee..aa8abdcab 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -152,6 +152,7 @@ public class Chunk {
         this.neighbors &= ~(0x1 << (x * 5 + 12 + z));
     }
     // CraftBukkit end
+    public final long chunkKey;
 
     public Chunk(World world, int i, int j) {
         this.sections = new ChunkSection[16];
@@ -165,6 +166,7 @@ public class Chunk {
         this.world = world;
         this.locX = i;
         this.locZ = j;
+        this.chunkKey = org.bukkit.craftbukkit.util.LongHash.toLong(i, j); // Blossom
         this.heightMap = new int[256];
 
         for (int k = 0; k < this.entitySlices.length; ++k) {
diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 243e3bfa1..d0f55a7ff 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -38,7 +38,7 @@ public class ChunkProviderServer implements IChunkProvider {
     public IChunkProvider chunkProvider;
     private IChunkLoader chunkLoader;
     public boolean forceChunkLoad = false; // CraftBukkit - true -> false
-    public Long2ObjectMap<Chunk> chunks = new Long2ObjectOpenHashMap<Chunk>(4096, 0.5f); // TacoSpigot - use trove Long2ObjectOpenHashMap instead of craftbukkit implementation (using inital capacity and load factor chosen by Amaranth in an old impl)
+    public Long2ObjectMap<Chunk> chunks = new me.samsuik.blossom.chunk.ServerChunkMap(4096, 0.5f); // Blossom // TacoSpigot - use trove Long2ObjectOpenHashMap instead of craftbukkit implementation (using inital capacity and load factor chosen by Amaranth in an old impl)
     public WorldServer world;
 
     public ChunkProviderServer(WorldServer worldserver, IChunkLoader ichunkloader, IChunkProvider ichunkprovider) {
-- 
2.50.1

