From 7c2399e3378fc2fe134966c34e693ebd4228bcb6 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 14 Jul 2023 22:31:07 +0100
Subject: [PATCH] Disable Player Suffocation Teleport


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index ef343897e..e4bc960f0 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -484,4 +484,9 @@ public class BlossomConfig {
         playerBoundPotions = getBoolean("projectiles.potion.player-bound-potions", true);
     }
 
+    public static boolean playerSuffocationTeleport;
+    private static void PlayerSuffocationTeleport() {
+        playerSuffocationTeleport = getBoolean("player.teleport-up-to-avoid-suffocation", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 58487b1b3..d9897fb30 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -82,7 +82,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     public int movementTick = 0; // Blossom
     // Blossom start - bind entities to player ticking
     public final List<Entity> boundEntities = new ArrayList<>();
-    
+
     public void tickBoundEntities() {
         for (int i = 0; i < boundEntities.size(); ++i) {
             Entity entity = boundEntities.get(i);
@@ -138,6 +138,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.S = 0.0F;
         this.setPositionRotation(blockposition, 0.0F, 0.0F);
 
+        if (me.samsuik.blossom.configuration.BlossomConfig.playerSuffocationTeleport) // Blossom
         while (!worldserver.getCubes(this, this.getBoundingBox()).isEmpty() && this.locY < 255.0D) {
             this.setPosition(this.locX, this.locY + 1.0D, this.locZ);
         }
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index d69a4731f..0ffb3277c 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -600,6 +600,7 @@ public abstract class PlayerList {
 
         worldserver.chunkProviderServer.getChunkAt((int) entityplayer1.locX >> 4, (int) entityplayer1.locZ >> 4);
 
+        if (me.samsuik.blossom.configuration.BlossomConfig.playerSuffocationTeleport) // Blossom
         while (avoidSuffocation && !worldserver.getCubes(entityplayer1, entityplayer1.getBoundingBox()).isEmpty() && entityplayer1.locY < 256.0D) {
             entityplayer1.setPosition(entityplayer1.locX, entityplayer1.locY + 1.0D, entityplayer1.locZ);
         }
-- 
2.40.0.windows.1

