From d8824a92b7f942c951be5ed3930b0bead4683978 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 18 Jul 2023 23:32:01 +0100
Subject: [PATCH] Actual Throttle Placement Configuration

smh, paper

diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 6a7e22160..83f3beb71 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -565,4 +565,9 @@ public class BlossomConfig {
         chunkLoadingPlayersPerThread = getInt("chunk.loading.players-per-thread", 50);
     }
 
+    public static boolean throttlePlacement;
+    private static void ThrottlePlacement() {
+        throttlePlacement = getBoolean("player.throttle-block-placement", false);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 8b40bec5d..f5ba9141b 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -915,6 +915,11 @@ public class PlayerConnection implements PacketListenerPlayIn, IUpdatePlayerList
             this.lastInteractedSlot = this.player.inventory.itemInHandIndex;
         }
         // Blossom end
+        // Blossom start
+        if (!me.samsuik.blossom.configuration.BlossomConfig.throttlePlacement && itemstack != null && !itemstack.hasTag()) {
+            lastPlace = -1;
+        }
+        // Blossom end
 
         boolean flag = false;
         BlockPosition blockposition = packetplayinblockplace.a();
-- 
2.50.1

