From 0961967c15b9da8a5f75836c86e5203f5b2ac86c Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 15 Jul 2023 14:31:04 +0100
Subject: [PATCH] Ensure interacted block is visible


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index d8a60ccff..016de39db 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -494,4 +494,9 @@ public class BlossomConfig {
         playerSuffocationTeleport = getBoolean("player.teleport-up-to-avoid-suffocation", false);
     }
 
+    public static boolean ensureInteractedBlockIsVisible;
+    private static void EnsureInteractedBlockIsVisible() {
+        ensureInteractedBlockIsVisible = getBoolean("player.ensure-interacted-block-is-visible", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/ItemMonsterEgg.java b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
index 26667abf8..506bffc40 100644
--- a/src/main/java/net/minecraft/server/ItemMonsterEgg.java
+++ b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
@@ -43,6 +43,14 @@ public class ItemMonsterEgg extends Item {
                 }
             }
 
+            // Blossom start
+            if (me.samsuik.blossom.configuration.BlossomConfig.ensureInteractedBlockIsVisible
+                && PlayerInteractManager.isObscured(blockposition, enumdirection, entityhuman, world)
+            ) {
+                return false;
+            }
+            // Blossom end
+
             blockposition = blockposition.shift(enumdirection);
             double d0 = 0.0D;
 
diff --git a/src/main/java/net/minecraft/server/PlayerInteractManager.java b/src/main/java/net/minecraft/server/PlayerInteractManager.java
index 9b1aeb5ee..971b5a29e 100644
--- a/src/main/java/net/minecraft/server/PlayerInteractManager.java
+++ b/src/main/java/net/minecraft/server/PlayerInteractManager.java
@@ -406,6 +406,14 @@ public class PlayerInteractManager {
         }
     }
 
+    // Blossom start
+    public static boolean isObscured(BlockPosition position, EnumDirection direction, EntityHuman entityhuman, World world) {
+        BlockPosition shifted = position.shift(direction);
+        Vec3D point = new Vec3D(shifted.getX() + 0.5, shifted.getY() + 0.5, shifted.getZ() + 0.5);
+        return world.a(point, entityhuman.getBoundingBox()) == 0.0f;
+    }
+    // Blossom end
+
     // CraftBukkit start
     public boolean interactResult = false;
     public boolean firedInteract = false;
@@ -505,6 +513,13 @@ public class PlayerInteractManager {
 
                 return false;
             } else if (!entityhuman.isSneaking() || itemstack == null) {
+                // Blossom start
+                Block block = blockdata.getBlock();
+                if (!me.samsuik.blossom.configuration.BlossomConfig.ensureInteractedBlockIsVisible
+                    || !(block instanceof IContainer)
+                    || !isObscured(blockposition, enumdirection, entityhuman, world)
+                )
+                // Blossom end
                 result = blockdata.getBlock().interact(world, blockposition, blockdata, entityhuman, enumdirection, f, f1, f2);
             }
 
-- 
2.40.0.windows.1

