From b54c7d92649343755375ac0841f9de99b27b2027 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 18 Jul 2023 20:13:30 +0100
Subject: [PATCH] Configure Lava Block Generation


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index 9465b33af..ccc0db646 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -531,4 +531,13 @@ public class BlossomConfig {
         disallowSuspiciousMovements = getBoolean("player.disallow-suspicious-movement", true);
     }
 
+    public static boolean lavaStoneGeneration;
+    public static boolean lavaCobbleGeneration;
+    public static boolean lavaObsidianGeneration;
+    private static void LavaGeneration() {
+        lavaStoneGeneration = getBoolean("fluids.lava.generate-stone", true);
+        lavaCobbleGeneration = getBoolean("fluids.lava.generate-cobblestone", true);
+        lavaObsidianGeneration = getBoolean("fluids.lava.create-obsidian", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index 7b48815ee..2948ec77c 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -111,7 +111,7 @@ public class BlockFlowing extends BlockFluids {
                 server.getPluginManager().callEvent(event);
             }
             if (!event.isCancelled()) {
-            if (this.material == Material.LAVA && world.getType(blockposition.down()).getBlock().getMaterial() == Material.WATER) {
+            if (this.material == Material.LAVA && world.getType(blockposition.down()).getBlock().getMaterial() == Material.WATER && me.samsuik.blossom.configuration.BlossomConfig.lavaStoneGeneration) { // Blossom
                 world.setTypeUpdate(blockposition.down(), Blocks.STONE.getBlockData());
                 this.fizz(world, blockposition.down());
                 return;
diff --git a/src/main/java/net/minecraft/server/BlockFluids.java b/src/main/java/net/minecraft/server/BlockFluids.java
index 8447dc78e..190b5b7e6 100644
--- a/src/main/java/net/minecraft/server/BlockFluids.java
+++ b/src/main/java/net/minecraft/server/BlockFluids.java
@@ -148,13 +148,13 @@ public abstract class BlockFluids extends Block {
             if (flag) {
                 Integer integer = (Integer) iblockdata.get(BlockFluids.LEVEL);
 
-                if (integer.intValue() == 0) {
+                if (integer.intValue() == 0 && me.samsuik.blossom.configuration.BlossomConfig.lavaObsidianGeneration) { // Blossom
                     world.setTypeUpdate(blockposition, Blocks.OBSIDIAN.getBlockData());
                     this.fizz(world, blockposition);
                     return true;
                 }
 
-                if (integer.intValue() > 0) { // PaperSpigot
+                if (integer.intValue() > 0 && me.samsuik.blossom.configuration.BlossomConfig.lavaCobbleGeneration) { // Blossom // PaperSpigot
                     world.setTypeUpdate(blockposition, Blocks.COBBLESTONE.getBlockData());
                     this.fizz(world, blockposition);
                     return true;
-- 
2.50.1

