From fe623573611133aa66e27cfbe6ab30df14711641 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Sat, 15 Jul 2023 17:48:22 +0100
Subject: [PATCH] Chunk Unload Delay


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index be7b7e319..de8e02f04 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -503,4 +503,9 @@ public class BlossomConfig {
         chunkCacheSize = getInt("chunk.chunk-cache-size", 8);
     }
 
+    public static int delayChunkUnloads;
+    private static void DelayChunkUnload() {
+        delayChunkUnloads = getInt("chunk.chunk-unload-delay", 100);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 5af16974b..99187971c 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -153,6 +153,7 @@ public class Chunk {
     }
     // CraftBukkit end
     public final long chunkKey;
+    public long loadedTick; // Blossom
 
     public Chunk(World world, int i, int j) {
         this.sections = new ChunkSection[16];
diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 3821a9cf7..1701ea0e6 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -148,6 +148,7 @@ public class ChunkProviderServer implements IChunkProvider {
             runnable.run();
         }
 
+        chunk.loadedTick = MinecraftServer.currentTick; // Blossom
         return chunk;
     }
     public Chunk originalGetChunkAt(int i, int j) {
@@ -374,9 +375,15 @@ public class ChunkProviderServer implements IChunkProvider {
             LongIterator iterator = unloadQueue.iterator();
             for (int i = 0; i < 100 && iterator.hasNext(); ++i) {
                 long chunkcoordinates = iterator.next();
-                iterator.remove();
+                // Blossom start
                 // TacoSpigot end
                 Chunk chunk = this.chunks.get(chunkcoordinates);
+                if (chunk != null && MinecraftServer.currentTick - chunk.loadedTick < me.samsuik.blossom.configuration.BlossomConfig.delayChunkUnloads) {
+                    i--;
+                    continue;
+                }
+                iterator.remove();
+                // Blossom end
                 if (chunk == null) continue;
 
                 ChunkUnloadEvent event = new ChunkUnloadEvent(chunk.bukkitChunk);
-- 
2.40.0.windows.1

