From f1eaf1fc0db643d88b2cce7be493137227fd7019 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 18 Jul 2023 20:20:39 +0100
Subject: [PATCH] Waterproof Redstone


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index ccc0db646..82ca6664f 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -540,4 +540,9 @@ public class BlossomConfig {
         lavaObsidianGeneration = getBoolean("fluids.lava.create-obsidian", true);
     }
 
+    public static boolean waterProofRedstone;
+    private static void WaterProofRedstone() {
+        waterProofRedstone = getBoolean("fluids.water-proof-redstone", true);
+    }
+
 }
diff --git a/src/main/java/net/minecraft/server/BlockFlowing.java b/src/main/java/net/minecraft/server/BlockFlowing.java
index 2948ec77c..7d2caaaff 100644
--- a/src/main/java/net/minecraft/server/BlockFlowing.java
+++ b/src/main/java/net/minecraft/server/BlockFlowing.java
@@ -236,6 +236,11 @@ public class BlockFlowing extends BlockFluids {
 
     private boolean g(World world, BlockPosition blockposition, IBlockData iblockdata) {
         Block block = world.getType(blockposition).getBlock();
+        // Blossom start - waterproof redstone
+        if (me.samsuik.blossom.configuration.BlossomConfig.waterProofRedstone) {
+            return !(block == Blocks.AIR || block instanceof BlockFluids);
+        }
+        // Blossom end - waterproof redstone
 
         return !(block instanceof BlockDoor) && block != Blocks.STANDING_SIGN && block != Blocks.LADDER && block != Blocks.REEDS ? (block.material == Material.PORTAL ? true : block.material.isSolid()) : true;
     }
-- 
2.50.1

