From dd35357354b731e2dc46235b7e4b47cf9fa5eebc Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Tue, 18 Jul 2023 22:04:29 +0100
Subject: [PATCH] Configure Chunk Executor Threads


diff --git a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
index f363ee9f5..c4c1b5b5a 100644
--- a/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
+++ b/src/main/java/me/samsuik/blossom/configuration/BlossomConfig.java
@@ -538,5 +538,12 @@ public class BlossomConfig {
         lavaCobbleThreshold = getInt("fluids.cobble-generation-threshold", -1);
         if (lavaCobbleThreshold < 0) lavaCobbleThreshold = Integer.MAX_VALUE;
     }
+    
+    public static int chunkLoadingBaseThreads;
+    public static int chunkLoadingPlayersPerThread;
+    private static void ChunkLoadingThreads() {
+        chunkLoadingBaseThreads = getInt("chunk.loading.base-threads", 2);
+        chunkLoadingPlayersPerThread = getInt("chunk.loading.players-per-thread", 50);
+    }
 
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
index 69e5da4c6..79cd50023 100644
--- a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
+++ b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
@@ -7,8 +7,10 @@ import net.minecraft.server.World;
 import org.bukkit.craftbukkit.util.AsynchronousExecutor;
 
 public class ChunkIOExecutor {
-    static final int BASE_THREADS = 2; // PaperSpigot - Bumped value
-    static final int PLAYERS_PER_THREAD = 50;
+    // Blossom start
+    static final int BASE_THREADS = me.samsuik.blossom.configuration.BlossomConfig.chunkLoadingBaseThreads; // PaperSpigot - Bumped value
+    static final int PLAYERS_PER_THREAD = me.samsuik.blossom.configuration.BlossomConfig.chunkLoadingPlayersPerThread;
+    // Blossom end
 
     private static final AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException> instance = new AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException>(new ChunkIOProvider(), BASE_THREADS);
 
-- 
2.40.0.windows.1

