From ccad1f231b8dff5936f8fba89abc84e8b1269399 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Mon, 3 Aug 2026 20:28:14 +0100
Subject: [PATCH] Ensure the block event queue is fully drained


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 47ab74a86..e8114378c 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1196,7 +1196,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     }
 
     private void ak() {
-        while (!this.S[this.T].isEmpty()) {
+        while (!this.S[0].isEmpty() || !this.S[1].isEmpty()) { // Blossom - ensure the block event queue is fully drained
             int i = this.T;
 
             this.T ^= 1;
-- 
2.50.1 (Apple Git-155)

