From 00272607cdf123b0e9beb146a340021d0d71da68 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Mon, 11 Mar 2024 12:55:45 +0000
Subject: [PATCH] Use optimised collection for gamerules


diff --git a/src/main/java/net/minecraft/server/GameRules.java b/src/main/java/net/minecraft/server/GameRules.java
index 1c89c9058..763545d65 100644
--- a/src/main/java/net/minecraft/server/GameRules.java
+++ b/src/main/java/net/minecraft/server/GameRules.java
@@ -3,10 +3,15 @@ package net.minecraft.server;
 import java.util.Iterator;
 import java.util.Set;
 import java.util.TreeMap;
+// Blossom start
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+// Blossom end
 
 public class GameRules {
 
     private TreeMap<String, GameRules.GameRuleValue> a = new TreeMap();
+    private final Int2ObjectMap<GameRuleValue> gameRules = new Int2ObjectOpenHashMap<>(); // Blossom
 
     public GameRules() {
         this.a("doFireTick", "true", GameRules.EnumGameRuleType.BOOLEAN_VALUE);
@@ -28,10 +33,11 @@ public class GameRules {
 
     public void a(String s, String s1, GameRules.EnumGameRuleType gamerules_enumgameruletype) {
         this.a.put(s, new GameRules.GameRuleValue(s1, gamerules_enumgameruletype));
+        this.gameRules.put(s.hashCode(), this.a.get(s)); // Blossom
     }
 
     public void set(String s, String s1) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.gameRules.get(s.hashCode()); // Blossom
 
         if (gamerules_gamerulevalue != null) {
             gamerules_gamerulevalue.a(s1);
@@ -42,19 +48,19 @@ public class GameRules {
     }
 
     public String get(String s) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.gameRules.get(s.hashCode()); // Blossom
 
         return gamerules_gamerulevalue != null ? gamerules_gamerulevalue.a() : "";
     }
 
     public boolean getBoolean(String s) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.gameRules.get(s.hashCode()); // Blossom
 
         return gamerules_gamerulevalue != null ? gamerules_gamerulevalue.b() : false;
     }
 
     public int c(String s) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.gameRules.get(s.hashCode()); // Blossom
 
         return gamerules_gamerulevalue != null ? gamerules_gamerulevalue.c() : 0;
     }
@@ -65,7 +71,7 @@ public class GameRules {
 
         while (iterator.hasNext()) {
             String s = (String) iterator.next();
-            GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+            GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.gameRules.get(s.hashCode()); // Blossom
 
             nbttagcompound.setString(s, gamerules_gamerulevalue.a());
         }
@@ -97,7 +103,7 @@ public class GameRules {
     }
 
     public boolean a(String s, GameRules.EnumGameRuleType gamerules_enumgameruletype) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.gameRules.get(s.hashCode()); // Blossom
 
         return gamerules_gamerulevalue != null && (gamerules_gamerulevalue.e() == gamerules_enumgameruletype || gamerules_enumgameruletype == GameRules.EnumGameRuleType.ANY_VALUE);
     }
-- 
2.40.0.windows.1

