From a06be08c41143e4587865d9917f539f4b6e41800 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 19 Jul 2025 17:17:02 +0100
Subject: [PATCH] Fix items bouncing on top of hoppers


diff --git a/src/main/java/net/minecraft/server/BlockHopper.java b/src/main/java/net/minecraft/server/BlockHopper.java
index 1bfbea2d2..b78f033c9 100644
--- a/src/main/java/net/minecraft/server/BlockHopper.java
+++ b/src/main/java/net/minecraft/server/BlockHopper.java
@@ -28,6 +28,13 @@ public class BlockHopper extends BlockContainer {
     }
 
     public void a(World world, BlockPosition blockposition, IBlockData iblockdata, AxisAlignedBB axisalignedbb, List<AxisAlignedBB> list, Entity entity) {
+        // Blossom start - fix items bouncing on top of hoppers
+        if (entity instanceof EntityItem) {
+            this.updateShape(world, blockposition);
+            super.a(world, blockposition, iblockdata, axisalignedbb, list, entity);
+            return;
+        }
+        // Blossom end - fix items bouncing on top of hoppers
         this.a(0.0F, 0.0F, 0.0F, 1.0F, 0.625F, 1.0F);
         super.a(world, blockposition, iblockdata, axisalignedbb, list, entity);
         float f = 0.125F;
-- 
2.50.1

