From 5509f21155d73a4f71e9aa192015563ce7ec14a4 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Wed, 26 Jul 2023 23:23:57 +0100
Subject: [PATCH] Use array for gamerules


diff --git a/src/main/java/net/minecraft/server/GameRules.java b/src/main/java/net/minecraft/server/GameRules.java
index 1c89c9058..082f2e015 100644
--- a/src/main/java/net/minecraft/server/GameRules.java
+++ b/src/main/java/net/minecraft/server/GameRules.java
@@ -7,6 +7,19 @@ import java.util.TreeMap;
 public class GameRules {
 
     private TreeMap<String, GameRules.GameRuleValue> a = new TreeMap();
+    // Blossom start
+    private final GameRuleValue[] storage = new GameRuleValue[32];
+    private int counter = 0;
+
+    private GameRuleValue value(String name) {
+        return storage[index(name)];
+    }
+
+    private int index(String name) {
+        // todo: is the string hashcode cached?
+        return (name.charAt(8) ^ (name.charAt(4) * 31)) & 31;
+    }
+    // Blossom end
 
     public GameRules() {
         this.a("doFireTick", "true", GameRules.EnumGameRuleType.BOOLEAN_VALUE);
@@ -28,10 +41,12 @@ public class GameRules {
 
     public void a(String s, String s1, GameRules.EnumGameRuleType gamerules_enumgameruletype) {
         this.a.put(s, new GameRules.GameRuleValue(s1, gamerules_enumgameruletype));
+        storage[counter] = a.get(s); // Blossom
+        storage[counter].name = s; // Blossom
     }
 
     public void set(String s, String s1) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) value(s); // Blossom
 
         if (gamerules_gamerulevalue != null) {
             gamerules_gamerulevalue.a(s1);
@@ -42,31 +57,30 @@ public class GameRules {
     }
 
     public String get(String s) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) value(s); // Blossom
 
         return gamerules_gamerulevalue != null ? gamerules_gamerulevalue.a() : "";
     }
 
     public boolean getBoolean(String s) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) value(s); // Blossom
 
         return gamerules_gamerulevalue != null ? gamerules_gamerulevalue.b() : false;
     }
 
     public int c(String s) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) value(s); // Blossom
 
         return gamerules_gamerulevalue != null ? gamerules_gamerulevalue.c() : 0;
     }
 
     public NBTTagCompound a() {
         NBTTagCompound nbttagcompound = new NBTTagCompound();
-        Iterator iterator = this.a.keySet().iterator();
-
-        while (iterator.hasNext()) {
-            String s = (String) iterator.next();
-            GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
-
+        // Blossom start
+        for (GameRuleValue gamerules_gamerulevalue : storage) {
+            if (gamerules_gamerulevalue == null) continue;
+            String s = gamerules_gamerulevalue.name;
+            // Blossom end
             nbttagcompound.setString(s, gamerules_gamerulevalue.a());
         }
 
@@ -93,11 +107,11 @@ public class GameRules {
     }
 
     public boolean contains(String s) {
-        return this.a.containsKey(s);
+        return value(s) != null; // Blossom
     }
 
     public boolean a(String s, GameRules.EnumGameRuleType gamerules_enumgameruletype) {
-        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) this.a.get(s);
+        GameRules.GameRuleValue gamerules_gamerulevalue = (GameRules.GameRuleValue) value(s); // Blossom
 
         return gamerules_gamerulevalue != null && (gamerules_gamerulevalue.e() == gamerules_enumgameruletype || gamerules_enumgameruletype == GameRules.EnumGameRuleType.ANY_VALUE);
     }
@@ -116,6 +130,7 @@ public class GameRules {
         private int c;
         private double d;
         private final GameRules.EnumGameRuleType e;
+        String name; // Blossom
 
         public GameRuleValue(String s, GameRules.EnumGameRuleType gamerules_enumgameruletype) {
             this.e = gamerules_enumgameruletype;
-- 
2.40.0.windows.1

