From be509f2890bebac39e6b9aa2a4f96a44e824881a Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sun, 13 Jul 2025 18:37:24 +0100
Subject: [PATCH] Catch some exceptions in the tick loop


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 232c1d30d..f86846d77 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -231,7 +231,15 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         // CraftBukkit end
         timings.doChunkUnload.startTiming(); // Spigot
         this.methodProfiler.c("chunkSource");
-        this.chunkProvider.unloadChunks();
+        // Blossom start - catch more exceptions in the world
+        try {
+            this.chunkProvider.unloadChunks();
+        } catch (RuntimeException rte) {
+            throw rte;
+        } catch (Exception ex) {
+            a.warn("An exception occurred while unloading chunks", ex);
+        }
+        // Blossom end - catch more exceptions in the world
         int j = this.a(1.0F);
 
         if (j != this.ab()) {
@@ -246,7 +254,17 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         timings.doChunkUnload.stopTiming(); // Spigot
         this.methodProfiler.c("tickPending");
         timings.scheduledBlocks.startTiming(); // Spigot
-        this.a(false);
+        // Blossom start - catch more exceptions in the world
+        try {
+            this.a(false);
+        } catch (RuntimeException rte) {
+            throw rte;
+        } catch (Exception ex) {
+            a.warn("An exception occurred while ticking blocks", ex);
+            M.clear();
+            V.clear();
+        }
+        // Blossom end - catch more exceptions in the world
         timings.scheduledBlocks.stopTiming(); // Spigot
         this.methodProfiler.c("tickBlocks");
         timings.chunkTicks.startTiming(); // Spigot
-- 
2.50.1

